<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 12 Cell type annotation | Orchestrating Single-Cell Analysis with Bioconductor</title>
  <meta name="description" content="Or: how I learned to stop worrying and love the t-SNEs." />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 12 Cell type annotation | Orchestrating Single-Cell Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="https://www.bioconductor.org/images/logo_bioconductor.gif" />
  <meta property="og:description" content="Or: how I learned to stop worrying and love the t-SNEs." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 12 Cell type annotation | Orchestrating Single-Cell Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Or: how I learned to stop worrying and love the t-SNEs." />
  <meta name="twitter:image" content="https://www.bioconductor.org/images/logo_bioconductor.gif" />



<meta name="date" content="2020-08-18" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="marker-detection.html"/>
<link rel="next" href="integrating-datasets.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Preamble</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#what-you-will-learn"><i class="fa fa-check"></i><b>1.1</b> What you will learn</a><ul>
<li class="chapter" data-level="1.1.1" data-path="introduction.html"><a href="introduction.html#preliminaries"><i class="fa fa-check"></i><b>1.1.1</b> Preliminaries</a></li>
<li class="chapter" data-level="1.1.2" data-path="introduction.html"><a href="introduction.html#workflows"><i class="fa fa-check"></i><b>1.1.2</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#what-you-wont-learn"><i class="fa fa-check"></i><b>1.2</b> What you won’t learn</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#who-we-wrote-this-for"><i class="fa fa-check"></i><b>1.3</b> Who we wrote this for</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#why-we-wrote-this"><i class="fa fa-check"></i><b>1.4</b> Why we wrote this</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.5</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html"><i class="fa fa-check"></i><b>2</b> Learning R and Bioconductor</a><ul>
<li class="chapter" data-level="2.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#the-benefits-of-r-and-bioconductor"><i class="fa fa-check"></i><b>2.1</b> The Benefits of R and Bioconductor</a></li>
<li class="chapter" data-level="2.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.2</b> Learning R Online</a></li>
<li class="chapter" data-level="2.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#running-r-locally"><i class="fa fa-check"></i><b>2.3</b> Running R Locally</a><ul>
<li class="chapter" data-level="2.3.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r"><i class="fa fa-check"></i><b>2.3.1</b> Installing R</a></li>
<li class="chapter" data-level="2.3.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#installing-r-bioconductor-packages"><i class="fa fa-check"></i><b>2.3.2</b> Installing R &amp; Bioconductor Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#getting-help-in-and-out-of-r"><i class="fa fa-check"></i><b>2.4</b> Getting Help In (and Out) of R</a></li>
<li class="chapter" data-level="2.5" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-documentation"><i class="fa fa-check"></i><b>2.5</b> Bioconductor Help</a><ul>
<li class="chapter" data-level="2.5.1" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.1</b> Bioconductor Packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#biocviews"><i class="fa fa-check"></i><b>2.5.2</b> biocViews</a></li>
<li class="chapter" data-level="2.5.3" data-path="learning-r-and-more.html"><a href="learning-r-and-more.html#bioconductor-forums"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor Forums</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html"><i class="fa fa-check"></i><b>3</b> Beyond R Basics</a><ul>
<li class="chapter" data-level="3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-an-r-expert"><i class="fa fa-check"></i><b>3.1</b> Becoming an R Expert</a></li>
<li class="chapter" data-level="3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#becoming-a-bioconductor-developer"><i class="fa fa-check"></i><b>3.2</b> Becoming an R/Bioconductor Developer</a></li>
<li class="chapter" data-level="3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#nice-companions-for-r"><i class="fa fa-check"></i><b>3.3</b> Nice Companions for R</a><ul>
<li class="chapter" data-level="3.3.1" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#shellbash"><i class="fa fa-check"></i><b>3.3.1</b> Shell/Bash</a></li>
<li class="chapter" data-level="3.3.2" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#git"><i class="fa fa-check"></i><b>3.3.2</b> Git</a></li>
<li class="chapter" data-level="3.3.3" data-path="beyond-r-basics.html"><a href="beyond-r-basics.html#other-languages"><i class="fa fa-check"></i><b>3.3.3</b> Other Languages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-infrastructure.html"><a href="data-infrastructure.html"><i class="fa fa-check"></i><b>4</b> Data Infrastructure</a><ul>
<li class="chapter" data-level="4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background"><i class="fa fa-check"></i><b>4.1</b> Background</a></li>
<li class="chapter" data-level="4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#storing-primary-experimental-data"><i class="fa fa-check"></i><b>4.2</b> Storing primary experimental data</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#filling-the-assays-slot"><i class="fa fa-check"></i><b>4.2.1</b> Filling the <code>assays</code> slot</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#adding-more-assays"><i class="fa fa-check"></i><b>4.2.2</b> Adding more <code>assays</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#handling-metadata"><i class="fa fa-check"></i><b>4.3</b> Handling metadata</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#on-the-columns"><i class="fa fa-check"></i><b>4.3.1</b> On the columns</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#on-the-rows"><i class="fa fa-check"></i><b>4.3.2</b> On the rows</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#other-metadata"><i class="fa fa-check"></i><b>4.3.3</b> Other metadata</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#single-cell-specific-fields"><i class="fa fa-check"></i><b>4.4</b> Single-cell-specific fields</a><ul>
<li class="chapter" data-level="4.4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background-1"><i class="fa fa-check"></i><b>4.4.1</b> Background</a></li>
<li class="chapter" data-level="4.4.2" data-path="data-infrastructure.html"><a href="data-infrastructure.html#dimensionality-reduction-results"><i class="fa fa-check"></i><b>4.4.2</b> Dimensionality reduction results</a></li>
<li class="chapter" data-level="4.4.3" data-path="data-infrastructure.html"><a href="data-infrastructure.html#alternative-experiments"><i class="fa fa-check"></i><b>4.4.3</b> Alternative Experiments</a></li>
<li class="chapter" data-level="4.4.4" data-path="data-infrastructure.html"><a href="data-infrastructure.html#size-factors"><i class="fa fa-check"></i><b>4.4.4</b> Size factors</a></li>
<li class="chapter" data-level="4.4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#column-labels"><i class="fa fa-check"></i><b>4.4.5</b> Column labels</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="data-infrastructure.html"><a href="data-infrastructure.html#conclusion"><i class="fa fa-check"></i><b>4.5</b> Conclusion</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="5" data-path="overview.html"><a href="overview.html"><i class="fa fa-check"></i><b>5</b> Overview</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="overview.html"><a href="overview.html#experimental-design"><i class="fa fa-check"></i><b>5.2</b> Experimental Design</a></li>
<li class="chapter" data-level="5.3" data-path="overview.html"><a href="overview.html#obtaining-a-count-matrix"><i class="fa fa-check"></i><b>5.3</b> Obtaining a count matrix</a></li>
<li class="chapter" data-level="5.4" data-path="overview.html"><a href="overview.html#data-processing-and-downstream-analysis"><i class="fa fa-check"></i><b>5.4</b> Data processing and downstream analysis</a></li>
<li class="chapter" data-level="5.5" data-path="overview.html"><a href="overview.html#quick-start"><i class="fa fa-check"></i><b>5.5</b> Quick start</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>6</b> Quality Control</a><ul>
<li class="chapter" data-level="6.1" data-path="quality-control.html"><a href="quality-control.html#quality-control-motivation"><i class="fa fa-check"></i><b>6.1</b> Motivation</a></li>
<li class="chapter" data-level="6.2" data-path="quality-control.html"><a href="quality-control.html#choice-of-qc-metrics"><i class="fa fa-check"></i><b>6.2</b> Choice of QC metrics</a></li>
<li class="chapter" data-level="6.3" data-path="quality-control.html"><a href="quality-control.html#identifying-low-quality-cells"><i class="fa fa-check"></i><b>6.3</b> Identifying low-quality cells</a><ul>
<li class="chapter" data-level="6.3.1" data-path="quality-control.html"><a href="quality-control.html#fixed-qc"><i class="fa fa-check"></i><b>6.3.1</b> With fixed thresholds</a></li>
<li class="chapter" data-level="6.3.2" data-path="quality-control.html"><a href="quality-control.html#quality-control-outlier"><i class="fa fa-check"></i><b>6.3.2</b> With adaptive thresholds</a></li>
<li class="chapter" data-level="6.3.3" data-path="quality-control.html"><a href="quality-control.html#other-approaches"><i class="fa fa-check"></i><b>6.3.3</b> Other approaches</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="quality-control.html"><a href="quality-control.html#quality-control-plots"><i class="fa fa-check"></i><b>6.4</b> Checking diagnostic plots</a></li>
<li class="chapter" data-level="6.5" data-path="quality-control.html"><a href="quality-control.html#quality-control-discarded"><i class="fa fa-check"></i><b>6.5</b> Removing low-quality cells</a></li>
<li class="chapter" data-level="6.6" data-path="quality-control.html"><a href="quality-control.html#marking-qc"><i class="fa fa-check"></i><b>6.6</b> Marking low-quality cells</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>7</b> Normalization</a><ul>
<li class="chapter" data-level="7.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>7.1</b> Motivation</a></li>
<li class="chapter" data-level="7.2" data-path="normalization.html"><a href="normalization.html#library-size-normalization"><i class="fa fa-check"></i><b>7.2</b> Library size normalization</a></li>
<li class="chapter" data-level="7.3" data-path="normalization.html"><a href="normalization.html#normalization-by-deconvolution"><i class="fa fa-check"></i><b>7.3</b> Normalization by deconvolution</a></li>
<li class="chapter" data-level="7.4" data-path="normalization.html"><a href="normalization.html#spike-norm"><i class="fa fa-check"></i><b>7.4</b> Normalization by spike-ins</a></li>
<li class="chapter" data-level="7.5" data-path="normalization.html"><a href="normalization.html#applying-the-size-factors"><i class="fa fa-check"></i><b>7.5</b> Applying the size factors</a><ul>
<li class="chapter" data-level="7.5.1" data-path="normalization.html"><a href="normalization.html#normalization-transformation"><i class="fa fa-check"></i><b>7.5.1</b> Scaling and log-transforming</a></li>
<li class="chapter" data-level="7.5.2" data-path="normalization.html"><a href="normalization.html#downsampling-and-log-transforming"><i class="fa fa-check"></i><b>7.5.2</b> Downsampling and log-transforming</a></li>
<li class="chapter" data-level="7.5.3" data-path="normalization.html"><a href="normalization.html#other-options"><i class="fa fa-check"></i><b>7.5.3</b> Other options</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>8</b> Feature selection</a><ul>
<li class="chapter" data-level="8.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>8.1</b> Motivation</a></li>
<li class="chapter" data-level="8.2" data-path="feature-selection.html"><a href="feature-selection.html#quantifying-per-gene-variation"><i class="fa fa-check"></i><b>8.2</b> Quantifying per-gene variation</a><ul>
<li class="chapter" data-level="8.2.1" data-path="feature-selection.html"><a href="feature-selection.html#variance-of-the-log-counts"><i class="fa fa-check"></i><b>8.2.1</b> Variance of the log-counts</a></li>
<li class="chapter" data-level="8.2.2" data-path="feature-selection.html"><a href="feature-selection.html#coefficient-of-variation"><i class="fa fa-check"></i><b>8.2.2</b> Coefficient of variation</a></li>
<li class="chapter" data-level="8.2.3" data-path="feature-selection.html"><a href="feature-selection.html#sec:spikeins"><i class="fa fa-check"></i><b>8.2.3</b> Quantifying technical noise</a></li>
<li class="chapter" data-level="8.2.4" data-path="feature-selection.html"><a href="feature-selection.html#accounting-for-blocking-factors"><i class="fa fa-check"></i><b>8.2.4</b> Accounting for blocking factors</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="feature-selection.html"><a href="feature-selection.html#hvg-selection"><i class="fa fa-check"></i><b>8.3</b> Selecting highly variable genes</a><ul>
<li class="chapter" data-level="8.3.1" data-path="overview.html"><a href="overview.html#overview"><i class="fa fa-check"></i><b>8.3.1</b> Overview</a></li>
<li class="chapter" data-level="8.3.2" data-path="feature-selection.html"><a href="feature-selection.html#based-on-the-largest-metrics"><i class="fa fa-check"></i><b>8.3.2</b> Based on the largest metrics</a></li>
<li class="chapter" data-level="8.3.3" data-path="feature-selection.html"><a href="feature-selection.html#based-on-significance"><i class="fa fa-check"></i><b>8.3.3</b> Based on significance</a></li>
<li class="chapter" data-level="8.3.4" data-path="feature-selection.html"><a href="feature-selection.html#feature-selection-positive"><i class="fa fa-check"></i><b>8.3.4</b> Keeping all genes above the trend</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="feature-selection.html"><a href="feature-selection.html#apriori-hvgs"><i class="fa fa-check"></i><b>8.4</b> Selecting <em>a priori</em> genes of interest</a></li>
<li class="chapter" data-level="8.5" data-path="feature-selection.html"><a href="feature-selection.html#feature-selection-subsetting"><i class="fa fa-check"></i><b>8.5</b> Putting it all together</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>9</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="9.1" data-path="overview.html"><a href="overview.html#overview"><i class="fa fa-check"></i><b>9.1</b> Overview</a></li>
<li class="chapter" data-level="9.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#principal-components-analysis"><i class="fa fa-check"></i><b>9.2</b> Principal components analysis</a></li>
<li class="chapter" data-level="9.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#choosing-the-number-of-pcs"><i class="fa fa-check"></i><b>9.3</b> Choosing the number of PCs</a><ul>
<li class="chapter" data-level="9.3.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>9.3.1</b> Motivation</a></li>
<li class="chapter" data-level="9.3.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#using-the-elbow-point"><i class="fa fa-check"></i><b>9.3.2</b> Using the elbow point</a></li>
<li class="chapter" data-level="9.3.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#using-the-technical-noise"><i class="fa fa-check"></i><b>9.3.3</b> Using the technical noise</a></li>
<li class="chapter" data-level="9.3.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#based-on-population-structure"><i class="fa fa-check"></i><b>9.3.4</b> Based on population structure</a></li>
<li class="chapter" data-level="9.3.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#using-random-matrix-theory"><i class="fa fa-check"></i><b>9.3.5</b> Using random matrix theory</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction-for-visualization"><i class="fa fa-check"></i><b>9.4</b> Dimensionality reduction for visualization</a><ul>
<li class="chapter" data-level="9.4.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#motivation-1"><i class="fa fa-check"></i><b>9.4.1</b> Motivation</a></li>
<li class="chapter" data-level="9.4.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#visualizing-with-pca"><i class="fa fa-check"></i><b>9.4.2</b> Visualizing with PCA</a></li>
<li class="chapter" data-level="9.4.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#t-stochastic-neighbor-embedding"><i class="fa fa-check"></i><b>9.4.3</b> <span class="math inline">\(t\)</span>-stochastic neighbor embedding</a></li>
<li class="chapter" data-level="9.4.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#uniform-manifold-approximation-and-projection"><i class="fa fa-check"></i><b>9.4.4</b> Uniform manifold approximation and projection</a></li>
<li class="chapter" data-level="9.4.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#visualization-interpretation"><i class="fa fa-check"></i><b>9.4.5</b> Interpreting the plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Clustering</a><ul>
<li class="chapter" data-level="10.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>10.1</b> Motivation</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#what-is-the-true-clustering"><i class="fa fa-check"></i><b>10.2</b> What is the “true clustering”?</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#clustering-graph"><i class="fa fa-check"></i><b>10.3</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="10.3.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background"><i class="fa fa-check"></i><b>10.3.1</b> Background</a></li>
<li class="chapter" data-level="10.3.2" data-path="clustering.html"><a href="clustering.html#implementation"><i class="fa fa-check"></i><b>10.3.2</b> Implementation</a></li>
<li class="chapter" data-level="10.3.3" data-path="clustering.html"><a href="clustering.html#other-parameters"><i class="fa fa-check"></i><b>10.3.3</b> Other parameters</a></li>
<li class="chapter" data-level="10.3.4" data-path="clustering.html"><a href="clustering.html#assessing-cluster-separation"><i class="fa fa-check"></i><b>10.3.4</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#k-means-clustering"><i class="fa fa-check"></i><b>10.4</b> <span class="math inline">\(k\)</span>-means clustering</a><ul>
<li class="chapter" data-level="10.4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background-1"><i class="fa fa-check"></i><b>10.4.1</b> Background</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#base-implementation"><i class="fa fa-check"></i><b>10.4.2</b> Base implementation</a></li>
<li class="chapter" data-level="10.4.3" data-path="clustering.html"><a href="clustering.html#assessing-cluster-separation-1"><i class="fa fa-check"></i><b>10.4.3</b> Assessing cluster separation</a></li>
<li class="chapter" data-level="10.4.4" data-path="clustering.html"><a href="clustering.html#in-two-step-procedures"><i class="fa fa-check"></i><b>10.4.4</b> In two-step procedures</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.5</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="10.5.1" data-path="clustering.html"><a href="clustering.html#background-2"><i class="fa fa-check"></i><b>10.5.1</b> Background</a></li>
<li class="chapter" data-level="10.5.2" data-path="clustering.html"><a href="clustering.html#implementation-1"><i class="fa fa-check"></i><b>10.5.2</b> Implementation</a></li>
<li class="chapter" data-level="10.5.3" data-path="clustering.html"><a href="clustering.html#silhouette-width"><i class="fa fa-check"></i><b>10.5.3</b> Assessing cluster separation</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="clustering.html"><a href="clustering.html#general-purpose-cluster-diagnostics"><i class="fa fa-check"></i><b>10.6</b> General-purpose cluster diagnostics</a><ul>
<li class="chapter" data-level="10.6.1" data-path="clustering.html"><a href="clustering.html#cluster-separation-redux"><i class="fa fa-check"></i><b>10.6.1</b> Cluster separation, redux</a></li>
<li class="chapter" data-level="10.6.2" data-path="clustering.html"><a href="clustering.html#comparing-different-clusterings"><i class="fa fa-check"></i><b>10.6.2</b> Comparing different clusterings</a></li>
<li class="chapter" data-level="10.6.3" data-path="clustering.html"><a href="clustering.html#cluster-bootstrapping"><i class="fa fa-check"></i><b>10.6.3</b> Evaluating cluster stability</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="clustering.html"><a href="clustering.html#subclustering"><i class="fa fa-check"></i><b>10.7</b> Subclustering</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="marker-detection.html"><a href="marker-detection.html"><i class="fa fa-check"></i><b>11</b> Marker gene detection</a><ul>
<li class="chapter" data-level="11.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>11.1</b> Motivation</a></li>
<li class="chapter" data-level="11.2" data-path="marker-detection.html"><a href="marker-detection.html#pairwise-tests-between-clusters"><i class="fa fa-check"></i><b>11.2</b> Pairwise tests between clusters</a><ul>
<li class="chapter" data-level="11.2.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#motivation-1"><i class="fa fa-check"></i><b>11.2.1</b> Motivation</a></li>
<li class="chapter" data-level="11.2.2" data-path="marker-detection.html"><a href="marker-detection.html#combining-pairwise-statistics-per-cluster"><i class="fa fa-check"></i><b>11.2.2</b> Combining pairwise statistics per cluster</a></li>
<li class="chapter" data-level="11.2.3" data-path="marker-detection.html"><a href="marker-detection.html#using-the-log-fold-change"><i class="fa fa-check"></i><b>11.2.3</b> Using the log-fold change</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="marker-detection.html"><a href="marker-detection.html#marker-tests"><i class="fa fa-check"></i><b>11.3</b> Alternative testing regimes</a><ul>
<li class="chapter" data-level="11.3.1" data-path="marker-detection.html"><a href="marker-detection.html#using-the-wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>11.3.1</b> Using the Wilcoxon rank sum test</a></li>
<li class="chapter" data-level="11.3.2" data-path="marker-detection.html"><a href="marker-detection.html#using-a-binomial-test"><i class="fa fa-check"></i><b>11.3.2</b> Using a binomial test</a></li>
<li class="chapter" data-level="11.3.3" data-path="marker-detection.html"><a href="marker-detection.html#using-custom-de-methods"><i class="fa fa-check"></i><b>11.3.3</b> Using custom DE methods</a></li>
<li class="chapter" data-level="11.3.4" data-path="marker-detection.html"><a href="marker-detection.html#combining-multiple-marker-statistics"><i class="fa fa-check"></i><b>11.3.4</b> Combining multiple marker statistics</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="marker-detection.html"><a href="marker-detection.html#marker-batch"><i class="fa fa-check"></i><b>11.4</b> Handling blocking factors</a><ul>
<li class="chapter" data-level="11.4.1" data-path="marker-detection.html"><a href="marker-detection.html#using-the-block-argument"><i class="fa fa-check"></i><b>11.4.1</b> Using the <code>block=</code> argument</a></li>
<li class="chapter" data-level="11.4.2" data-path="marker-detection.html"><a href="marker-detection.html#using-the-design-argument"><i class="fa fa-check"></i><b>11.4.2</b> Using the <code>design=</code> argument</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="marker-detection.html"><a href="marker-detection.html#p-value-invalidity"><i class="fa fa-check"></i><b>11.5</b> Invalidity of <span class="math inline">\(p\)</span>-values</a><ul>
<li class="chapter" data-level="11.5.1" data-path="marker-detection.html"><a href="marker-detection.html#from-data-snooping"><i class="fa fa-check"></i><b>11.5.1</b> From data snooping</a></li>
<li class="chapter" data-level="11.5.2" data-path="marker-detection.html"><a href="marker-detection.html#false-replicates"><i class="fa fa-check"></i><b>11.5.2</b> Nature of replication</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="marker-detection.html"><a href="marker-detection.html#further-comments"><i class="fa fa-check"></i><b>11.6</b> Further comments</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html"><i class="fa fa-check"></i><b>12</b> Cell type annotation</a><ul>
<li class="chapter" data-level="12.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>12.1</b> Motivation</a></li>
<li class="chapter" data-level="12.2" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#assigning-cell-labels-from-reference-data"><i class="fa fa-check"></i><b>12.2</b> Assigning cell labels from reference data</a><ul>
<li class="chapter" data-level="12.2.1" data-path="overview.html"><a href="overview.html#overview"><i class="fa fa-check"></i><b>12.2.1</b> Overview</a></li>
<li class="chapter" data-level="12.2.2" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#using-existing-references"><i class="fa fa-check"></i><b>12.2.2</b> Using existing references</a></li>
<li class="chapter" data-level="12.2.3" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#using-custom-references"><i class="fa fa-check"></i><b>12.2.3</b> Using custom references</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#assigning-cell-labels-from-gene-sets"><i class="fa fa-check"></i><b>12.3</b> Assigning cell labels from gene sets</a></li>
<li class="chapter" data-level="12.4" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#assigning-cluster-labels-from-markers"><i class="fa fa-check"></i><b>12.4</b> Assigning cluster labels from markers</a></li>
<li class="chapter" data-level="12.5" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#computing-gene-set-activities"><i class="fa fa-check"></i><b>12.5</b> Computing gene set activities</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="integrating-datasets.html"><a href="integrating-datasets.html"><i class="fa fa-check"></i><b>13</b> Integrating Datasets</a><ul>
<li class="chapter" data-level="13.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>13.1</b> Motivation</a></li>
<li class="chapter" data-level="13.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#setting-up-the-data"><i class="fa fa-check"></i><b>13.2</b> Setting up the data</a></li>
<li class="chapter" data-level="13.3" data-path="integrating-datasets.html"><a href="integrating-datasets.html#batch-diagnosis"><i class="fa fa-check"></i><b>13.3</b> Diagnosing batch effects</a></li>
<li class="chapter" data-level="13.4" data-path="integrating-datasets.html"><a href="integrating-datasets.html#linear-regression"><i class="fa fa-check"></i><b>13.4</b> Linear regression</a></li>
<li class="chapter" data-level="13.5" data-path="integrating-datasets.html"><a href="integrating-datasets.html#performing-mnn-correction"><i class="fa fa-check"></i><b>13.5</b> Performing MNN correction</a><ul>
<li class="chapter" data-level="13.5.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#algorithm-overview"><i class="fa fa-check"></i><b>13.5.1</b> Algorithm overview</a></li>
<li class="chapter" data-level="13.5.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#application-to-the-pbmc-data"><i class="fa fa-check"></i><b>13.5.2</b> Application to the PBMC data</a></li>
<li class="chapter" data-level="13.5.3" data-path="integrating-datasets.html"><a href="integrating-datasets.html#correction-diagnostics"><i class="fa fa-check"></i><b>13.5.3</b> Correction diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="13.6" data-path="integrating-datasets.html"><a href="integrating-datasets.html#preserving-biological-heterogeneity"><i class="fa fa-check"></i><b>13.6</b> Preserving biological heterogeneity</a><ul>
<li class="chapter" data-level="13.6.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#comparison-to-within-batch-clusters"><i class="fa fa-check"></i><b>13.6.1</b> Comparison to within-batch clusters</a></li>
<li class="chapter" data-level="13.6.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#integration-with-markers"><i class="fa fa-check"></i><b>13.6.2</b> Encouraging consistency with marker genes</a></li>
</ul></li>
<li class="chapter" data-level="13.7" data-path="integrating-datasets.html"><a href="integrating-datasets.html#using-corrected-values"><i class="fa fa-check"></i><b>13.7</b> Using the corrected values</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html"><i class="fa fa-check"></i><b>14</b> Multi-sample comparisons</a><ul>
<li class="chapter" data-level="14.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>14.1</b> Motivation</a></li>
<li class="chapter" data-level="14.2" data-path="integrating-datasets.html"><a href="integrating-datasets.html#setting-up-the-data"><i class="fa fa-check"></i><b>14.2</b> Setting up the data</a></li>
<li class="chapter" data-level="14.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#differential-expression-between-conditions"><i class="fa fa-check"></i><b>14.3</b> Differential expression between conditions</a><ul>
<li class="chapter" data-level="14.3.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#creating-pseudo-bulk-samples"><i class="fa fa-check"></i><b>14.3.1</b> Creating pseudo-bulk samples</a></li>
<li class="chapter" data-level="14.3.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#performing-the-de-analysis"><i class="fa fa-check"></i><b>14.3.2</b> Performing the DE analysis</a></li>
<li class="chapter" data-level="14.3.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#putting-it-all-together"><i class="fa fa-check"></i><b>14.3.3</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#ambient-problems"><i class="fa fa-check"></i><b>14.4</b> Avoiding problems with ambient RNA</a><ul>
<li class="chapter" data-level="14.4.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#motivation-1"><i class="fa fa-check"></i><b>14.4.1</b> Motivation</a></li>
<li class="chapter" data-level="14.4.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#finding-affected-degs"><i class="fa fa-check"></i><b>14.4.2</b> Finding affected DEGs</a></li>
<li class="chapter" data-level="14.4.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#subtracting-ambient-counts"><i class="fa fa-check"></i><b>14.4.3</b> Subtracting ambient counts</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#differential-abundance"><i class="fa fa-check"></i><b>14.5</b> Differential abundance between conditions</a><ul>
<li class="chapter" data-level="14.5.1" data-path="overview.html"><a href="overview.html#overview"><i class="fa fa-check"></i><b>14.5.1</b> Overview</a></li>
<li class="chapter" data-level="14.5.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#performing-the-da-analysis"><i class="fa fa-check"></i><b>14.5.2</b> Performing the DA analysis</a></li>
<li class="chapter" data-level="14.5.3" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#composition-effects"><i class="fa fa-check"></i><b>14.5.3</b> Handling composition effects</a></li>
</ul></li>
<li class="chapter" data-level="14.6" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#comments-on-interpretation"><i class="fa fa-check"></i><b>14.6</b> Comments on interpretation</a><ul>
<li class="chapter" data-level="14.6.1" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#de-da-duality"><i class="fa fa-check"></i><b>14.6.1</b> DE or DA? Two sides of the same coin</a></li>
<li class="chapter" data-level="14.6.2" data-path="multi-sample-comparisons.html"><a href="multi-sample-comparisons.html#sacrificing-differences"><i class="fa fa-check"></i><b>14.6.2</b> Sacrificing biology by integration</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="droplet-processing.html"><a href="droplet-processing.html"><i class="fa fa-check"></i><b>15</b> Droplet processing</a><ul>
<li class="chapter" data-level="15.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>15.1</b> Motivation</a></li>
<li class="chapter" data-level="15.2" data-path="droplet-processing.html"><a href="droplet-processing.html#qc-droplets"><i class="fa fa-check"></i><b>15.2</b> Calling cells from empty droplets</a><ul>
<li class="chapter" data-level="15.2.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background"><i class="fa fa-check"></i><b>15.2.1</b> Background</a></li>
<li class="chapter" data-level="15.2.2" data-path="droplet-processing.html"><a href="droplet-processing.html#testing-for-empty-droplets"><i class="fa fa-check"></i><b>15.2.2</b> Testing for empty droplets</a></li>
<li class="chapter" data-level="15.2.3" data-path="droplet-processing.html"><a href="droplet-processing.html#relationship-with-other-qc-metrics"><i class="fa fa-check"></i><b>15.2.3</b> Relationship with other QC metrics</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="droplet-processing.html"><a href="droplet-processing.html#cell-hashing"><i class="fa fa-check"></i><b>15.3</b> Demultiplexing cell hashes</a><ul>
<li class="chapter" data-level="15.3.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background-1"><i class="fa fa-check"></i><b>15.3.1</b> Background</a></li>
<li class="chapter" data-level="15.3.2" data-path="droplet-processing.html"><a href="droplet-processing.html#cell-calling-options"><i class="fa fa-check"></i><b>15.3.2</b> Cell calling options</a></li>
<li class="chapter" data-level="15.3.3" data-path="droplet-processing.html"><a href="droplet-processing.html#demultiplexing-on-hto-abundance"><i class="fa fa-check"></i><b>15.3.3</b> Demultiplexing on HTO abundance</a></li>
</ul></li>
<li class="chapter" data-level="15.4" data-path="marker-detection.html"><a href="marker-detection.html#further-comments"><i class="fa fa-check"></i><b>15.4</b> Further comments</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="doublet-detection.html"><a href="doublet-detection.html"><i class="fa fa-check"></i><b>16</b> Doublet detection</a><ul>
<li class="chapter" data-level="16.1" data-path="overview.html"><a href="overview.html#overview"><i class="fa fa-check"></i><b>16.1</b> Overview</a></li>
<li class="chapter" data-level="16.2" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-detection-with-clusters"><i class="fa fa-check"></i><b>16.2</b> Doublet detection with clusters</a></li>
<li class="chapter" data-level="16.3" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-simulation"><i class="fa fa-check"></i><b>16.3</b> Doublet detection by simulation</a></li>
<li class="chapter" data-level="16.4" data-path="doublet-detection.html"><a href="doublet-detection.html#doublet-detection-in-multiplexed-experiments"><i class="fa fa-check"></i><b>16.4</b> Doublet detection in multiplexed experiments</a><ul>
<li class="chapter" data-level="16.4.1" data-path="data-infrastructure.html"><a href="data-infrastructure.html#background"><i class="fa fa-check"></i><b>16.4.1</b> Background</a></li>
<li class="chapter" data-level="16.4.2" data-path="doublet-detection.html"><a href="doublet-detection.html#identifying-inter-sample-doublets"><i class="fa fa-check"></i><b>16.4.2</b> Identifying inter-sample doublets</a></li>
<li class="chapter" data-level="16.4.3" data-path="doublet-detection.html"><a href="doublet-detection.html#guilt-by-association-for-unmarked-doublets"><i class="fa fa-check"></i><b>16.4.3</b> Guilt by association for unmarked doublets</a></li>
</ul></li>
<li class="chapter" data-level="16.5" data-path="marker-detection.html"><a href="marker-detection.html#further-comments"><i class="fa fa-check"></i><b>16.5</b> Further comments</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html"><i class="fa fa-check"></i><b>17</b> Cell cycle assignment</a><ul>
<li class="chapter" data-level="17.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>17.1</b> Motivation</a></li>
<li class="chapter" data-level="17.2" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#using-the-cyclins"><i class="fa fa-check"></i><b>17.2</b> Using the cyclins</a></li>
<li class="chapter" data-level="17.3" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#using-reference-profiles"><i class="fa fa-check"></i><b>17.3</b> Using reference profiles</a></li>
<li class="chapter" data-level="17.4" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#using-the-cyclone-classifier"><i class="fa fa-check"></i><b>17.4</b> Using the <code>cyclone()</code> classifier</a></li>
<li class="chapter" data-level="17.5" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#removing-cell-cycle-effects"><i class="fa fa-check"></i><b>17.5</b> Removing cell cycle effects</a><ul>
<li class="chapter" data-level="17.5.1" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#comments"><i class="fa fa-check"></i><b>17.5.1</b> Comments</a></li>
<li class="chapter" data-level="17.5.2" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#with-linear-regression-and-friends"><i class="fa fa-check"></i><b>17.5.2</b> With linear regression and friends</a></li>
<li class="chapter" data-level="17.5.3" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#removing-cell-cycle-related-genes"><i class="fa fa-check"></i><b>17.5.3</b> Removing cell cycle-related genes</a></li>
<li class="chapter" data-level="17.5.4" data-path="cell-cycle-assignment.html"><a href="cell-cycle-assignment.html#using-contrastive-pca"><i class="fa fa-check"></i><b>17.5.4</b> Using contrastive PCA</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html"><i class="fa fa-check"></i><b>18</b> Trajectory Analysis</a><ul>
<li class="chapter" data-level="18.1" data-path="overview.html"><a href="overview.html#overview"><i class="fa fa-check"></i><b>18.1</b> Overview</a></li>
<li class="chapter" data-level="18.2" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#obtaining-pseudotime-orderings"><i class="fa fa-check"></i><b>18.2</b> Obtaining pseudotime orderings</a><ul>
<li class="chapter" data-level="18.2.1" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#overview-1"><i class="fa fa-check"></i><b>18.2.1</b> Overview</a></li>
<li class="chapter" data-level="18.2.2" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#cluster-based-minimum-spanning-tree"><i class="fa fa-check"></i><b>18.2.2</b> Cluster-based minimum spanning tree</a></li>
<li class="chapter" data-level="18.2.3" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#principal-curves"><i class="fa fa-check"></i><b>18.2.3</b> Principal curves</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#characterizing-trajectories"><i class="fa fa-check"></i><b>18.3</b> Characterizing trajectories</a><ul>
<li class="chapter" data-level="18.3.1" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#overview-2"><i class="fa fa-check"></i><b>18.3.1</b> Overview</a></li>
<li class="chapter" data-level="18.3.2" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#changes-along-a-trajectory"><i class="fa fa-check"></i><b>18.3.2</b> Changes along a trajectory</a></li>
<li class="chapter" data-level="18.3.3" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#changes-between-paths"><i class="fa fa-check"></i><b>18.3.3</b> Changes between paths</a></li>
<li class="chapter" data-level="18.3.4" data-path="marker-detection.html"><a href="marker-detection.html#further-comments"><i class="fa fa-check"></i><b>18.3.4</b> Further comments</a></li>
</ul></li>
<li class="chapter" data-level="18.4" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#finding-the-root"><i class="fa fa-check"></i><b>18.4</b> Finding the root</a><ul>
<li class="chapter" data-level="18.4.1" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#overview-3"><i class="fa fa-check"></i><b>18.4.1</b> Overview</a></li>
<li class="chapter" data-level="18.4.2" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#entropy-based-methods"><i class="fa fa-check"></i><b>18.4.2</b> Entropy-based methods</a></li>
<li class="chapter" data-level="18.4.3" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#rna-velocity"><i class="fa fa-check"></i><b>18.4.3</b> RNA velocity</a></li>
<li class="chapter" data-level="18.4.4" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#real-timepoints"><i class="fa fa-check"></i><b>18.4.4</b> Real timepoints</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="trajectory-analysis.html"><a href="trajectory-analysis.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="single-nuclei-rna-seq-processing.html"><a href="single-nuclei-rna-seq-processing.html"><i class="fa fa-check"></i><b>19</b> Single-nuclei RNA-seq processing</a><ul>
<li class="chapter" data-level="19.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>19.1</b> Introduction</a></li>
<li class="chapter" data-level="19.2" data-path="single-nuclei-rna-seq-processing.html"><a href="single-nuclei-rna-seq-processing.html#quality-control-for-stripped-nuclei"><i class="fa fa-check"></i><b>19.2</b> Quality control for stripped nuclei</a></li>
<li class="chapter" data-level="19.3" data-path="single-nuclei-rna-seq-processing.html"><a href="single-nuclei-rna-seq-processing.html#comments-on-downstream-analyses"><i class="fa fa-check"></i><b>19.3</b> Comments on downstream analyses</a></li>
<li class="chapter" data-level="19.4" data-path="single-nuclei-rna-seq-processing.html"><a href="single-nuclei-rna-seq-processing.html#tricks-with-ambient-contamination"><i class="fa fa-check"></i><b>19.4</b> Tricks with ambient contamination</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html"><i class="fa fa-check"></i><b>20</b> Integrating with protein abundance</a><ul>
<li class="chapter" data-level="20.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>20.1</b> Motivation</a></li>
<li class="chapter" data-level="20.2" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#preprocessing"><i class="fa fa-check"></i><b>20.2</b> Preprocessing</a><ul>
<li class="chapter" data-level="20.2.1" data-path="integrating-datasets.html"><a href="integrating-datasets.html#setting-up-the-data"><i class="fa fa-check"></i><b>20.2.1</b> Setting up the data</a></li>
<li class="chapter" data-level="20.2.2" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>20.2.2</b> Quality control</a></li>
<li class="chapter" data-level="20.2.3" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>20.2.3</b> Normalization</a></li>
</ul></li>
<li class="chapter" data-level="20.3" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#clustering-and-interpretation"><i class="fa fa-check"></i><b>20.3</b> Clustering and interpretation</a></li>
<li class="chapter" data-level="20.4" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#integration-with-gene-expression-data"><i class="fa fa-check"></i><b>20.4</b> Integration with gene expression data</a><ul>
<li class="chapter" data-level="20.4.1" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#by-subclustering"><i class="fa fa-check"></i><b>20.4.1</b> By subclustering</a></li>
<li class="chapter" data-level="20.4.2" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#by-combined-clustering"><i class="fa fa-check"></i><b>20.4.2</b> By combined clustering</a></li>
<li class="chapter" data-level="20.4.3" data-path="integrating-with-protein-abundance.html"><a href="integrating-with-protein-abundance.html#by-differential-testing"><i class="fa fa-check"></i><b>20.4.3</b> By differential testing</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="repertoire-seq.html"><a href="repertoire-seq.html"><i class="fa fa-check"></i><b>21</b> Analyzing repertoire sequencing data</a><ul>
<li class="chapter" data-level="21.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>21.1</b> Motivation</a></li>
<li class="chapter" data-level="21.2" data-path="repertoire-seq.html"><a href="repertoire-seq.html#loading-the-tcr-repertoire"><i class="fa fa-check"></i><b>21.2</b> Loading the TCR repertoire</a></li>
<li class="chapter" data-level="21.3" data-path="repertoire-seq.html"><a href="repertoire-seq.html#leveraging-list-semantics"><i class="fa fa-check"></i><b>21.3</b> Leveraging <code>List</code> semantics</a></li>
<li class="chapter" data-level="21.4" data-path="repertoire-seq.html"><a href="repertoire-seq.html#converting-back-to-dataframes"><i class="fa fa-check"></i><b>21.4</b> Converting back to <code>DataFrame</code>s</a></li>
<li class="chapter" data-level="21.5" data-path="repertoire-seq.html"><a href="repertoire-seq.html#case-study-for-clonotype-analyses"><i class="fa fa-check"></i><b>21.5</b> Case study for clonotype analyses</a></li>
<li class="chapter" data-level="21.6" data-path="repertoire-seq.html"><a href="repertoire-seq.html#repeating-for-immunoglobulins"><i class="fa fa-check"></i><b>21.6</b> Repeating for immunoglobulins</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="interactive-sharing.html"><a href="interactive-sharing.html"><i class="fa fa-check"></i><b>22</b> Interactive data exploration</a><ul>
<li class="chapter" data-level="22.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>22.1</b> Motivation</a></li>
<li class="chapter" data-level="22.2" data-path="interactive-sharing.html"><a href="interactive-sharing.html#interactive-quickstart"><i class="fa fa-check"></i><b>22.2</b> Quick start</a></li>
<li class="chapter" data-level="22.3" data-path="interactive-sharing.html"><a href="interactive-sharing.html#isee-examples"><i class="fa fa-check"></i><b>22.3</b> Usage examples</a><ul>
<li class="chapter" data-level="22.3.1" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>22.3.1</b> Quality control</a></li>
<li class="chapter" data-level="22.3.2" data-path="interactive-sharing.html"><a href="interactive-sharing.html#annotation-of-cell-populations"><i class="fa fa-check"></i><b>22.3.2</b> Annotation of cell populations</a></li>
<li class="chapter" data-level="22.3.3" data-path="interactive-sharing.html"><a href="interactive-sharing.html#querying-features-of-interest"><i class="fa fa-check"></i><b>22.3.3</b> Querying features of interest</a></li>
</ul></li>
<li class="chapter" data-level="22.4" data-path="interactive-sharing.html"><a href="interactive-sharing.html#reproducible-visualizations"><i class="fa fa-check"></i><b>22.4</b> Reproducible visualizations</a></li>
<li class="chapter" data-level="22.5" data-path="interactive-sharing.html"><a href="interactive-sharing.html#dissemination"><i class="fa fa-check"></i><b>22.5</b> Dissemination of analysis results</a></li>
<li class="chapter" data-level="22.6" data-path="interactive-sharing.html"><a href="interactive-sharing.html#additional-resources"><i class="fa fa-check"></i><b>22.6</b> Additional resources</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html"><i class="fa fa-check"></i><b>23</b> Dealing with big data</a><ul>
<li class="chapter" data-level="23.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>23.1</b> Motivation</a></li>
<li class="chapter" data-level="23.2" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#fast-approximations"><i class="fa fa-check"></i><b>23.2</b> Fast approximations</a><ul>
<li class="chapter" data-level="23.2.1" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#nearest-neighbor-searching"><i class="fa fa-check"></i><b>23.2.1</b> Nearest neighbor searching</a></li>
<li class="chapter" data-level="23.2.2" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#big-data-svd"><i class="fa fa-check"></i><b>23.2.2</b> Singular value decomposition</a></li>
</ul></li>
<li class="chapter" data-level="23.3" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#parallelization"><i class="fa fa-check"></i><b>23.3</b> Parallelization</a></li>
<li class="chapter" data-level="23.4" data-path="dealing-with-big-data.html"><a href="dealing-with-big-data.html#out-of-memory-representations"><i class="fa fa-check"></i><b>23.4</b> Out of memory representations</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="interoperability.html"><a href="interoperability.html"><i class="fa fa-check"></i><b>24</b> Interoperability</a><ul>
<li class="chapter" data-level="24.1" data-path="normalization.html"><a href="normalization.html#motivation"><i class="fa fa-check"></i><b>24.1</b> Motivation</a></li>
<li class="chapter" data-level="24.2" data-path="interoperability.html"><a href="interoperability.html#interchanging-with-seurat"><i class="fa fa-check"></i><b>24.2</b> Interchanging with <em>Seurat</em></a></li>
<li class="chapter" data-level="24.3" data-path="interoperability.html"><a href="interoperability.html#interchanging-with-scanpy"><i class="fa fa-check"></i><b>24.3</b> Interchanging with <em>scanpy</em></a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="25" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html"><i class="fa fa-check"></i><b>25</b> Lun 416B cell line (Smart-seq2)</a><ul>
<li class="chapter" data-level="25.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>25.1</b> Introduction</a></li>
<li class="chapter" data-level="25.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>25.2</b> Data loading</a></li>
<li class="chapter" data-level="25.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>25.3</b> Quality control</a></li>
<li class="chapter" data-level="25.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>25.4</b> Normalization</a></li>
<li class="chapter" data-level="25.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>25.5</b> Variance modelling</a></li>
<li class="chapter" data-level="25.6" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#batch-correction"><i class="fa fa-check"></i><b>25.6</b> Batch correction</a></li>
<li class="chapter" data-level="25.7" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>25.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="25.8" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>25.8</b> Clustering</a></li>
<li class="chapter" data-level="25.9" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#interpretation"><i class="fa fa-check"></i><b>25.9</b> Interpretation</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="zeisel-mouse-brain-strt-seq.html"><a href="zeisel-mouse-brain-strt-seq.html"><i class="fa fa-check"></i><b>26</b> Zeisel mouse brain (STRT-Seq)</a><ul>
<li class="chapter" data-level="26.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>26.1</b> Introduction</a></li>
<li class="chapter" data-level="26.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>26.2</b> Data loading</a></li>
<li class="chapter" data-level="26.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>26.3</b> Quality control</a></li>
<li class="chapter" data-level="26.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>26.4</b> Normalization</a></li>
<li class="chapter" data-level="26.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>26.5</b> Variance modelling</a></li>
<li class="chapter" data-level="26.6" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>26.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="26.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>26.7</b> Clustering</a></li>
<li class="chapter" data-level="26.8" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#interpretation"><i class="fa fa-check"></i><b>26.8</b> Interpretation</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="unfiltered-human-pbmcs-10x-genomics.html"><a href="unfiltered-human-pbmcs-10x-genomics.html"><i class="fa fa-check"></i><b>27</b> Unfiltered human PBMCs (10X Genomics)</a><ul>
<li class="chapter" data-level="27.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>27.1</b> Introduction</a></li>
<li class="chapter" data-level="27.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>27.2</b> Data loading</a></li>
<li class="chapter" data-level="27.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>27.3</b> Quality control</a></li>
<li class="chapter" data-level="27.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>27.4</b> Normalization</a></li>
<li class="chapter" data-level="27.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>27.5</b> Variance modelling</a></li>
<li class="chapter" data-level="27.6" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>27.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="27.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>27.7</b> Clustering</a></li>
<li class="chapter" data-level="27.8" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#interpretation"><i class="fa fa-check"></i><b>27.8</b> Interpretation</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="28" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html"><i class="fa fa-check"></i><b>28</b> Filtered human PBMCs (10X Genomics)</a><ul>
<li class="chapter" data-level="28.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>28.1</b> Introduction</a></li>
<li class="chapter" data-level="28.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>28.2</b> Data loading</a></li>
<li class="chapter" data-level="28.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>28.3</b> Quality control</a></li>
<li class="chapter" data-level="28.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>28.4</b> Normalization</a></li>
<li class="chapter" data-level="28.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>28.5</b> Variance modelling</a></li>
<li class="chapter" data-level="28.6" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>28.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="28.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>28.7</b> Clustering</a></li>
<li class="chapter" data-level="28.8" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#data-integration"><i class="fa fa-check"></i><b>28.8</b> Data integration</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="29" data-path="human-pbmc-with-surface-proteins-10x-genomics.html"><a href="human-pbmc-with-surface-proteins-10x-genomics.html"><i class="fa fa-check"></i><b>29</b> Human PBMC with surface proteins (10X Genomics)</a><ul>
<li class="chapter" data-level="29.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>29.1</b> Introduction</a></li>
<li class="chapter" data-level="29.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>29.2</b> Data loading</a></li>
<li class="chapter" data-level="29.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>29.3</b> Quality control</a></li>
<li class="chapter" data-level="29.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>29.4</b> Normalization</a></li>
<li class="chapter" data-level="29.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>29.5</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="29.6" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>29.6</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="30" data-path="grun-human-pancreas-cel-seq2.html"><a href="grun-human-pancreas-cel-seq2.html"><i class="fa fa-check"></i><b>30</b> Grun human pancreas (CEL-seq2)</a><ul>
<li class="chapter" data-level="30.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>30.1</b> Introduction</a></li>
<li class="chapter" data-level="30.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>30.2</b> Data loading</a></li>
<li class="chapter" data-level="30.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>30.3</b> Quality control</a></li>
<li class="chapter" data-level="30.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>30.4</b> Normalization</a></li>
<li class="chapter" data-level="30.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>30.5</b> Variance modelling</a></li>
<li class="chapter" data-level="30.6" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#data-integration"><i class="fa fa-check"></i><b>30.6</b> Data integration</a></li>
<li class="chapter" data-level="30.7" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>30.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="30.8" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>30.8</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="31" data-path="muraro-human-pancreas-cel-seq.html"><a href="muraro-human-pancreas-cel-seq.html"><i class="fa fa-check"></i><b>31</b> Muraro human pancreas (CEL-seq)</a><ul>
<li class="chapter" data-level="31.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>31.1</b> Introduction</a></li>
<li class="chapter" data-level="31.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>31.2</b> Data loading</a></li>
<li class="chapter" data-level="31.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>31.3</b> Quality control</a></li>
<li class="chapter" data-level="31.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>31.4</b> Normalization</a></li>
<li class="chapter" data-level="31.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>31.5</b> Variance modelling</a></li>
<li class="chapter" data-level="31.6" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#data-integration"><i class="fa fa-check"></i><b>31.6</b> Data integration</a></li>
<li class="chapter" data-level="31.7" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>31.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="31.8" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>31.8</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="32" data-path="lawlor-human-pancreas-smarter.html"><a href="lawlor-human-pancreas-smarter.html"><i class="fa fa-check"></i><b>32</b> Lawlor human pancreas (SMARTer)</a><ul>
<li class="chapter" data-level="32.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>32.1</b> Introduction</a></li>
<li class="chapter" data-level="32.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>32.2</b> Data loading</a></li>
<li class="chapter" data-level="32.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>32.3</b> Quality control</a></li>
<li class="chapter" data-level="32.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>32.4</b> Normalization</a></li>
<li class="chapter" data-level="32.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>32.5</b> Variance modelling</a></li>
<li class="chapter" data-level="32.6" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>32.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="32.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>32.7</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="33" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html"><i class="fa fa-check"></i><b>33</b> Segerstolpe human pancreas (Smart-seq2)</a><ul>
<li class="chapter" data-level="33.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>33.1</b> Introduction</a></li>
<li class="chapter" data-level="33.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>33.2</b> Data loading</a></li>
<li class="chapter" data-level="33.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>33.3</b> Quality control</a></li>
<li class="chapter" data-level="33.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>33.4</b> Normalization</a></li>
<li class="chapter" data-level="33.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>33.5</b> Variance modelling</a></li>
<li class="chapter" data-level="33.6" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>33.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="33.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>33.7</b> Clustering</a></li>
<li class="chapter" data-level="33.8" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#data-integration"><i class="fa fa-check"></i><b>33.8</b> Data integration</a></li>
<li class="chapter" data-level="33.9" data-path="segerstolpe-human-pancreas-smart-seq2.html"><a href="segerstolpe-human-pancreas-smart-seq2.html#segerstolpe-comparison"><i class="fa fa-check"></i><b>33.9</b> Multi-sample comparisons</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="34" data-path="merged-pancreas.html"><a href="merged-pancreas.html"><i class="fa fa-check"></i><b>34</b> Merged human pancreas datasets</a><ul>
<li class="chapter" data-level="34.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>34.1</b> Introduction</a></li>
<li class="chapter" data-level="34.2" data-path="merged-pancreas.html"><a href="merged-pancreas.html#the-good"><i class="fa fa-check"></i><b>34.2</b> The good</a></li>
<li class="chapter" data-level="34.3" data-path="merged-pancreas.html"><a href="merged-pancreas.html#the-bad"><i class="fa fa-check"></i><b>34.3</b> The bad</a></li>
<li class="chapter" data-level="34.4" data-path="merged-pancreas.html"><a href="merged-pancreas.html#the-ugly"><i class="fa fa-check"></i><b>34.4</b> The ugly</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="35" data-path="merged-hcsc.html"><a href="merged-hcsc.html"><i class="fa fa-check"></i><b>35</b> Grun mouse HSC (CEL-seq)</a><ul>
<li class="chapter" data-level="35.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>35.1</b> Introduction</a></li>
<li class="chapter" data-level="35.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>35.2</b> Data loading</a></li>
<li class="chapter" data-level="35.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>35.3</b> Quality control</a></li>
<li class="chapter" data-level="35.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>35.4</b> Normalization</a></li>
<li class="chapter" data-level="35.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>35.5</b> Variance modelling</a></li>
<li class="chapter" data-level="35.6" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>35.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="35.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>35.7</b> Clustering</a></li>
<li class="chapter" data-level="35.8" data-path="merged-hcsc.html"><a href="merged-hcsc.html#marker-gene-detection"><i class="fa fa-check"></i><b>35.8</b> Marker gene detection</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="36" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html"><i class="fa fa-check"></i><b>36</b> Nestorowa mouse HSC (Smart-seq2)</a><ul>
<li class="chapter" data-level="36.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>36.1</b> Introduction</a></li>
<li class="chapter" data-level="36.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>36.2</b> Data loading</a></li>
<li class="chapter" data-level="36.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>36.3</b> Quality control</a></li>
<li class="chapter" data-level="36.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>36.4</b> Normalization</a></li>
<li class="chapter" data-level="36.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>36.5</b> Variance modelling</a></li>
<li class="chapter" data-level="36.6" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>36.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="36.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>36.7</b> Clustering</a></li>
<li class="chapter" data-level="36.8" data-path="merged-hcsc.html"><a href="merged-hcsc.html#marker-gene-detection"><i class="fa fa-check"></i><b>36.8</b> Marker gene detection</a></li>
<li class="chapter" data-level="36.9" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#cell-type-annotation"><i class="fa fa-check"></i><b>36.9</b> Cell type annotation</a></li>
<li class="chapter" data-level="36.10" data-path="nestorowa-mouse-hsc-smart-seq2.html"><a href="nestorowa-mouse-hsc-smart-seq2.html#miscellaneous-analyses"><i class="fa fa-check"></i><b>36.10</b> Miscellaneous analyses</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="37" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><i class="fa fa-check"></i><b>37</b> Pijuan-Sala chimeric mouse embryo (10X Genomics)</a><ul>
<li class="chapter" data-level="37.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>37.1</b> Introduction</a></li>
<li class="chapter" data-level="37.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>37.2</b> Data loading</a></li>
<li class="chapter" data-level="37.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>37.3</b> Quality control</a></li>
<li class="chapter" data-level="37.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>37.4</b> Normalization</a></li>
<li class="chapter" data-level="37.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>37.5</b> Variance modelling</a></li>
<li class="chapter" data-level="37.6" data-path="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html"><a href="pijuan-sala-chimeric-mouse-embryo-10x-genomics.html#merging"><i class="fa fa-check"></i><b>37.6</b> Merging</a></li>
<li class="chapter" data-level="37.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>37.7</b> Clustering</a></li>
<li class="chapter" data-level="37.8" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>37.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="38" data-path="bach-mouse-mammary-gland-10x-genomics.html"><a href="bach-mouse-mammary-gland-10x-genomics.html"><i class="fa fa-check"></i><b>38</b> Bach mouse mammary gland (10X Genomics)</a><ul>
<li class="chapter" data-level="38.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>38.1</b> Introduction</a></li>
<li class="chapter" data-level="38.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>38.2</b> Data loading</a></li>
<li class="chapter" data-level="38.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>38.3</b> Quality control</a></li>
<li class="chapter" data-level="38.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>38.4</b> Normalization</a></li>
<li class="chapter" data-level="38.5" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#variance-modelling"><i class="fa fa-check"></i><b>38.5</b> Variance modelling</a></li>
<li class="chapter" data-level="38.6" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>38.6</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="38.7" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>38.7</b> Clustering</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="39" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html"><i class="fa fa-check"></i><b>39</b> HCA human bone marrow (10X Genomics)</a><ul>
<li class="chapter" data-level="39.1" data-path="introduction.html"><a href="introduction.html#introduction"><i class="fa fa-check"></i><b>39.1</b> Introduction</a></li>
<li class="chapter" data-level="39.2" data-path="lun-416b-cell-line-smart-seq2.html"><a href="lun-416b-cell-line-smart-seq2.html#data-loading"><i class="fa fa-check"></i><b>39.2</b> Data loading</a></li>
<li class="chapter" data-level="39.3" data-path="quality-control.html"><a href="quality-control.html#quality-control"><i class="fa fa-check"></i><b>39.3</b> Quality control</a></li>
<li class="chapter" data-level="39.4" data-path="normalization.html"><a href="normalization.html#normalization"><i class="fa fa-check"></i><b>39.4</b> Normalization</a></li>
<li class="chapter" data-level="39.5" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#variance-modeling"><i class="fa fa-check"></i><b>39.5</b> Variance modeling</a></li>
<li class="chapter" data-level="39.6" data-path="filtered-human-pbmcs-10x-genomics.html"><a href="filtered-human-pbmcs-10x-genomics.html#data-integration"><i class="fa fa-check"></i><b>39.6</b> Data integration</a></li>
<li class="chapter" data-level="39.7" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#dimensionality-reduction"><i class="fa fa-check"></i><b>39.7</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="39.8" data-path="clustering.html"><a href="clustering.html#clustering"><i class="fa fa-check"></i><b>39.8</b> Clustering</a></li>
<li class="chapter" data-level="39.9" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#differential-expression"><i class="fa fa-check"></i><b>39.9</b> Differential expression</a></li>
<li class="chapter" data-level="39.10" data-path="hca-human-bone-marrow-10x-genomics.html"><a href="hca-human-bone-marrow-10x-genomics.html#cell-type-classification"><i class="fa fa-check"></i><b>39.10</b> Cell type classification</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="40" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>40</b> Contributors</a><ul>
<li><a href="contributors.html#aaron-lun-phd"><em>Aaron Lun, PhD</em></a></li>
<li><a href="contributors.html#robert-amezquita-phd"><em>Robert Amezquita, PhD</em></a></li>
<li><a href="contributors.html#stephanie-hicks-phd"><em>Stephanie Hicks, PhD</em></a></li>
<li><a href="contributors.html#raphael-gottardo-phd"><em>Raphael Gottardo, PhD</em></a></li>
<li class="chapter" data-level="" data-path="contributors.html"><a href="contributors.html#other-notable-entities"><i class="fa fa-check"></i>Other notable entities</a></li>
</ul></li>
<li class="chapter" data-level="41" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>41</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Single-Cell Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="cell-type-annotation" class="section level1">
<h1><span class="header-section-number">Chapter 12</span> Cell type annotation</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="motivation" class="section level2">
<h2><span class="header-section-number">12.1</span> Motivation</h2>
<p>The most challenging task in scRNA-seq data analysis is arguably the interpretation of the results.
Obtaining clusters of cells is fairly straightforward, but it is more difficult to determine what biological state is represented by each of those clusters.
Doing so requires us to bridge the gap between the current dataset and prior biological knowledge, and the latter is not always available in a consistent and quantitative manner.
Indeed, even the concept of a “cell type” is <a href="https://doi.org/10.1016/j.cels.2017.03.006">not clearly defined</a>, with most practitioners possessing a “I’ll know it when I see it” intuition that is not amenable to computational analysis.
As such, interpretation of scRNA-seq data is often manual and a common bottleneck in the analysis workflow.</p>
<p>To expedite this step, we can use various computational approaches that exploit prior information to assign meaning to an uncharacterized scRNA-seq dataset.
The most obvious sources of prior information are the curated gene sets associated with particular biological processes, e.g., from the Gene Ontology (GO) or the Kyoto Encyclopedia of Genes and Genomes (KEGG) collections.
Alternatively, we can directly compare our expression profiles to published reference datasets where each sample or cell has already been annotated with its putative biological state by domain experts.
Here, we will demonstrate both approaches with several different scRNA-seq datasets.</p>
</div>
<div id="assigning-cell-labels-from-reference-data" class="section level2">
<h2><span class="header-section-number">12.2</span> Assigning cell labels from reference data</h2>
<div id="overview" class="section level3">
<h3><span class="header-section-number">12.2.1</span> Overview</h3>
<p>A conceptually straightforward annotation approach is to compare the single-cell expression profiles with previously annotated reference datasets.
Labels can then be assigned to each cell in our uncharacterized test dataset based on the most similar reference sample(s), for some definition of “similar”.
This is a standard classification challenge that can be tackled by standard machine learning techniques such as random forests and support vector machines.
Any published and labelled RNA-seq dataset (bulk or single-cell) can be used as a reference, though its reliability depends greatly on the expertise of the original authors who assigned the labels in the first place.</p>
<p>In this section, we will demonstrate the use of the <em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> method <span class="citation">(Aran et al. <a href="#ref-aran2019reference">2019</a>)</span> for cell type annotation.
This method assigns labels to cells based on the reference samples with the highest Spearman rank correlations, using only the marker genes between pairs of labels to focus on the relevant differences between cell types.
It also performs a fine-tuning step for each cell where the correlations are recomputed with just the marker genes for the top-scoring labels.
This aims to resolve any ambiguity between those labels by removing noise from irrelevant markers for other labels.
Further details can be found in the <a href="https://ltla.github.io/SingleRBook"><em>SingleR</em> book</a> from which most of the examples here are derived.</p>
</div>
<div id="using-existing-references" class="section level3">
<h3><span class="header-section-number">12.2.2</span> Using existing references</h3>
<p>For demonstration purposes, we will use one of the 10X PBMC datasets as our test.
While we have already applied quality control, normalization and clustering for this dataset, this is not strictly necessary.
It is entirely possible to run <code>SingleR()</code> on the raw counts without any <em>a priori</em> quality control
and filter on the annotation results at one’s leisure - see the book for an explanation.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb374-1" data-line-number="1"><span class="co">#--- loading ---#</span></a>
<a class="sourceLine" id="cb374-2" data-line-number="2"><span class="kw">library</span>(BiocFileCache)</a>
<a class="sourceLine" id="cb374-3" data-line-number="3">bfc &lt;-<span class="st"> </span><span class="kw">BiocFileCache</span>(<span class="st">&quot;raw_data&quot;</span>, <span class="dt">ask =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb374-4" data-line-number="4">raw.path &lt;-<span class="st"> </span><span class="kw">bfcrpath</span>(bfc, <span class="kw">file.path</span>(<span class="st">&quot;http://cf.10xgenomics.com/samples&quot;</span>,</a>
<a class="sourceLine" id="cb374-5" data-line-number="5">    <span class="st">&quot;cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz&quot;</span>))</a>
<a class="sourceLine" id="cb374-6" data-line-number="6"><span class="kw">untar</span>(raw.path, <span class="dt">exdir=</span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k&quot;</span>))</a>
<a class="sourceLine" id="cb374-7" data-line-number="7"></a>
<a class="sourceLine" id="cb374-8" data-line-number="8"><span class="kw">library</span>(DropletUtils)</a>
<a class="sourceLine" id="cb374-9" data-line-number="9">fname &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;pbmc4k/raw_gene_bc_matrices/GRCh38&quot;</span>)</a>
<a class="sourceLine" id="cb374-10" data-line-number="10">sce.pbmc &lt;-<span class="st"> </span><span class="kw">read10xCounts</span>(fname, <span class="dt">col.names=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb374-11" data-line-number="11"></a>
<a class="sourceLine" id="cb374-12" data-line-number="12"><span class="co">#--- gene-annotation ---#</span></a>
<a class="sourceLine" id="cb374-13" data-line-number="13"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb374-14" data-line-number="14"><span class="kw">rownames</span>(sce.pbmc) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(</a>
<a class="sourceLine" id="cb374-15" data-line-number="15">    <span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, <span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>Symbol)</a>
<a class="sourceLine" id="cb374-16" data-line-number="16"></a>
<a class="sourceLine" id="cb374-17" data-line-number="17"><span class="kw">library</span>(EnsDb.Hsapiens.v86)</a>
<a class="sourceLine" id="cb374-18" data-line-number="18">location &lt;-<span class="st"> </span><span class="kw">mapIds</span>(EnsDb.Hsapiens.v86, <span class="dt">keys=</span><span class="kw">rowData</span>(sce.pbmc)<span class="op">$</span>ID, </a>
<a class="sourceLine" id="cb374-19" data-line-number="19">    <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>, <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb374-20" data-line-number="20"></a>
<a class="sourceLine" id="cb374-21" data-line-number="21"><span class="co">#--- cell-detection ---#</span></a>
<a class="sourceLine" id="cb374-22" data-line-number="22"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb374-23" data-line-number="23">e.out &lt;-<span class="st"> </span><span class="kw">emptyDrops</span>(<span class="kw">counts</span>(sce.pbmc))</a>
<a class="sourceLine" id="cb374-24" data-line-number="24">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="kw">which</span>(e.out<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.001</span>)]</a>
<a class="sourceLine" id="cb374-25" data-line-number="25"></a>
<a class="sourceLine" id="cb374-26" data-line-number="26"><span class="co">#--- quality-control ---#</span></a>
<a class="sourceLine" id="cb374-27" data-line-number="27">stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce.pbmc, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span><span class="kw">which</span>(location<span class="op">==</span><span class="st">&quot;MT&quot;</span>)))</a>
<a class="sourceLine" id="cb374-28" data-line-number="28">high.mito &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(stats<span class="op">$</span>subsets_Mito_percent, <span class="dt">type=</span><span class="st">&quot;higher&quot;</span>)</a>
<a class="sourceLine" id="cb374-29" data-line-number="29">sce.pbmc &lt;-<span class="st"> </span>sce.pbmc[,<span class="op">!</span>high.mito]</a>
<a class="sourceLine" id="cb374-30" data-line-number="30"></a>
<a class="sourceLine" id="cb374-31" data-line-number="31"><span class="co">#--- normalization ---#</span></a>
<a class="sourceLine" id="cb374-32" data-line-number="32"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb374-33" data-line-number="33"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb374-34" data-line-number="34">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb374-35" data-line-number="35">sce.pbmc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.pbmc, <span class="dt">cluster=</span>clusters)</a>
<a class="sourceLine" id="cb374-36" data-line-number="36">sce.pbmc &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb374-37" data-line-number="37"></a>
<a class="sourceLine" id="cb374-38" data-line-number="38"><span class="co">#--- variance-modelling ---#</span></a>
<a class="sourceLine" id="cb374-39" data-line-number="39"><span class="kw">set.seed</span>(<span class="dv">1001</span>)</a>
<a class="sourceLine" id="cb374-40" data-line-number="40">dec.pbmc &lt;-<span class="st"> </span><span class="kw">modelGeneVarByPoisson</span>(sce.pbmc)</a>
<a class="sourceLine" id="cb374-41" data-line-number="41">top.pbmc &lt;-<span class="st"> </span><span class="kw">getTopHVGs</span>(dec.pbmc, <span class="dt">prop=</span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb374-42" data-line-number="42"></a>
<a class="sourceLine" id="cb374-43" data-line-number="43"><span class="co">#--- dimensionality-reduction ---#</span></a>
<a class="sourceLine" id="cb374-44" data-line-number="44"><span class="kw">set.seed</span>(<span class="dv">10000</span>)</a>
<a class="sourceLine" id="cb374-45" data-line-number="45">sce.pbmc &lt;-<span class="st"> </span><span class="kw">denoisePCA</span>(sce.pbmc, <span class="dt">subset.row=</span>top.pbmc, <span class="dt">technical=</span>dec.pbmc)</a>
<a class="sourceLine" id="cb374-46" data-line-number="46"></a>
<a class="sourceLine" id="cb374-47" data-line-number="47"><span class="kw">set.seed</span>(<span class="dv">100000</span>)</a>
<a class="sourceLine" id="cb374-48" data-line-number="48">sce.pbmc &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce.pbmc, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb374-49" data-line-number="49"></a>
<a class="sourceLine" id="cb374-50" data-line-number="50"><span class="kw">set.seed</span>(<span class="dv">1000000</span>)</a>
<a class="sourceLine" id="cb374-51" data-line-number="51">sce.pbmc &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce.pbmc, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb374-52" data-line-number="52"></a>
<a class="sourceLine" id="cb374-53" data-line-number="53"><span class="co">#--- clustering ---#</span></a>
<a class="sourceLine" id="cb374-54" data-line-number="54">g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce.pbmc, <span class="dt">k=</span><span class="dv">10</span>, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb374-55" data-line-number="55">clust &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb374-56" data-line-number="56"><span class="kw">colLabels</span>(sce.pbmc) &lt;-<span class="st"> </span><span class="kw">factor</span>(clust)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb375"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb375-1" data-line-number="1">sce.pbmc</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 3985 
## metadata(1): Samples
## assays(2): counts logcounts
## rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B
## rowData names(2): ID Symbol
## colnames(3985): AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 ...
##   TTTGTCAGTTAAGACA-1 TTTGTCATCCCAAGAT-1
## colData names(4): Sample Barcode sizeFactor label
## reducedDimNames(3): PCA TSNE UMAP
## altExpNames(0):</code></pre>
<p>The <em><a href="https://bioconductor.org/packages/3.12/celldex">celldex</a></em> contains a number of curated reference datasets, mostly assembled from bulk RNA-seq or microarray data of sorted cell types.
These references are often good enough for most applications provided that they contain the cell types that are expected in the test population.
Here, we will use a reference constructed from Blueprint and ENCODE data <span class="citation">(Martens and Stunnenberg <a href="#ref-martens2013blueprint">2013</a>; The ENCODE Project Consortium <a href="#ref-encode2012integrated">2012</a>)</span>;
this is obtained by calling the <code>BlueprintEncode()</code> function to construct a <code>SummarizedExperiment</code> containing log-expression values with curated labels for each sample.</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb377-1" data-line-number="1"><span class="kw">library</span>(celldex)</a>
<a class="sourceLine" id="cb377-2" data-line-number="2">ref &lt;-<span class="st"> </span><span class="kw">BlueprintEncodeData</span>()</a>
<a class="sourceLine" id="cb377-3" data-line-number="3">ref</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 19859 259 
## metadata(0):
## assays(1): logcounts
## rownames(19859): TSPAN6 TNMD ... LINC00550 GIMAP1-GIMAP5
## rowData names(0):
## colnames(259): mature.neutrophil
##   CD14.positive..CD16.negative.classical.monocyte ...
##   epithelial.cell.of.umbilical.artery.1
##   dermis.lymphatic.vessel.endothelial.cell.1
## colData names(3): label.main label.fine label.ont</code></pre>
<p>We call the <code>SingleR()</code> function to annotate each of our PBMCs with the main cell type labels from the Blueprint/ENCODE reference.
This returns a <code>DataFrame</code> where each row corresponds to a cell in the test dataset and contains its label assignments.
Alternatively, we could use the labels in <code>ref$label.fine</code>, which provide more resolution at the cost of speed and increased ambiguity in the assignments.</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb379-1" data-line-number="1"><span class="kw">library</span>(SingleR)</a>
<a class="sourceLine" id="cb379-2" data-line-number="2">pred &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sce.pbmc, <span class="dt">ref=</span>ref, <span class="dt">labels=</span>ref<span class="op">$</span>label.main)</a>
<a class="sourceLine" id="cb379-3" data-line-number="3"><span class="kw">table</span>(pred<span class="op">$</span>labels)</a></code></pre></div>
<pre><code>## 
##      B-cells CD4+ T-cells CD8+ T-cells           DC  Eosinophils Erythrocytes 
##          549          773         1274            1            1            5 
##          HSC    Monocytes     NK cells 
##           14         1117          251</code></pre>
<p>We inspect the results using a heatmap of the per-cell and label scores (Figure <a href="cell-type-annotation.html#fig:singler-heat-pbmc">12.1</a>).
Ideally, each cell should exhibit a high score in one label relative to all of the others, indicating that the assignment to that label was unambiguous.
This is largely the case for monocytes and B cells, whereas we see more ambiguity between CD4<sup>+</sup> and CD8<sup>+</sup> T cells (and to a lesser extent, NK cells).</p>
<div class="sourceCode" id="cb381"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb381-1" data-line-number="1"><span class="kw">plotScoreHeatmap</span>(pred)</a></code></pre></div>
<div class="figure"><span id="fig:singler-heat-pbmc"></span>
<img src="cell-annotation_files/figure-html/singler-heat-pbmc-1.png" alt="Heatmap of the assignment score for each cell (column) and label (row). Scores are shown before any fine-tuning and are normalized to [0, 1] within each cell." width="672" />
<p class="caption">
Figure 12.1: Heatmap of the assignment score for each cell (column) and label (row). Scores are shown before any fine-tuning and are normalized to [0, 1] within each cell.
</p>
</div>
<p>We compare the assignments with the clustering results to determine the identity of each cluster.
Here, several clusters are nested within the monocyte and B cell labels (Figure <a href="cell-type-annotation.html#fig:singler-cluster">12.2</a>), indicating that the clustering represents finer subdivisions within the cell types.
Interestingly, our clustering does not effectively distinguish between CD4<sup>+</sup> and CD8<sup>+</sup> T cell labels.
This is probably due to the presence of other factors of heterogeneity within the T cell subpopulation (e.g., activation) that have a stronger influence on unsupervised methods than the <em>a priori</em> expected CD4<sup>+</sup>/CD8<sup>+</sup> distinction.</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb382-1" data-line-number="1">tab &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">Assigned=</span>pred<span class="op">$</span>pruned.labels, <span class="dt">Cluster=</span><span class="kw">colLabels</span>(sce.pbmc))</a>
<a class="sourceLine" id="cb382-2" data-line-number="2"></a>
<a class="sourceLine" id="cb382-3" data-line-number="3"><span class="co"># Adding a pseudo-count of 10 to avoid strong color jumps with just 1 cell.</span></a>
<a class="sourceLine" id="cb382-4" data-line-number="4"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb382-5" data-line-number="5"><span class="kw">pheatmap</span>(<span class="kw">log2</span>(tab<span class="op">+</span><span class="dv">10</span>), <span class="dt">color=</span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;blue&quot;</span>))(<span class="dv">101</span>))</a></code></pre></div>
<div class="figure"><span id="fig:singler-cluster"></span>
<img src="cell-annotation_files/figure-html/singler-cluster-1.png" alt="Heatmap of the distribution of cells across labels and clusters in the 10X PBMC dataset. Color scale is reported in the log~10~-number of cells for each cluster-label combination." width="672" />
<p class="caption">
Figure 12.2: Heatmap of the distribution of cells across labels and clusters in the 10X PBMC dataset. Color scale is reported in the log<sub>10</sub>-number of cells for each cluster-label combination.
</p>
</div>
<p>This episode highlights some of the differences between reference-based annotation and unsupervised clustering.
The former explicitly focuses on aspects of the data that are known to be interesting, simplifying the process of biological interpretation.
However, the cost is that the downstream analysis is restricted by the diversity and resolution of the available labels, a problem that is largely avoided by <em>de novo</em> identification of clusters.
We suggest applying both strategies to examine the agreement (or lack thereof) between reference label and cluster assignments.
Any inconsistencies are not necessarily problematic due to the conceptual differences between the two approaches;
indeed, one could use those discrepancies as the basis for further investigation to discover novel factors of variation in the data.</p>
</div>
<div id="using-custom-references" class="section level3">
<h3><span class="header-section-number">12.2.3</span> Using custom references</h3>
<p>We can also apply <em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> to single-cell reference datasets that are curated and supplied by the user.
This is most obviously useful when we have an existing dataset that was previously (manually) annotated
and we want to use that knowledge to annotate a new dataset in an automated manner.
To illustrate, we will use the <span class="citation">Muraro et al. (<a href="#ref-muraro2016singlecell">2016</a>)</span> human pancreas dataset as our reference.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb383-1" data-line-number="1"><span class="co">#--- loading ---#</span></a>
<a class="sourceLine" id="cb383-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb383-3" data-line-number="3">sce.muraro &lt;-<span class="st"> </span><span class="kw">MuraroPancreasData</span>()</a>
<a class="sourceLine" id="cb383-4" data-line-number="4"></a>
<a class="sourceLine" id="cb383-5" data-line-number="5"><span class="co">#--- gene-annotation ---#</span></a>
<a class="sourceLine" id="cb383-6" data-line-number="6"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb383-7" data-line-number="7">edb &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73881&quot;</span>]]</a>
<a class="sourceLine" id="cb383-8" data-line-number="8">gene.symb &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;__chr.*$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">rownames</span>(sce.muraro))</a>
<a class="sourceLine" id="cb383-9" data-line-number="9">gene.ids &lt;-<span class="st"> </span><span class="kw">mapIds</span>(edb, <span class="dt">keys=</span>gene.symb, </a>
<a class="sourceLine" id="cb383-10" data-line-number="10">    <span class="dt">keytype=</span><span class="st">&quot;SYMBOL&quot;</span>, <span class="dt">column=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb383-11" data-line-number="11"></a>
<a class="sourceLine" id="cb383-12" data-line-number="12"><span class="co"># Removing duplicated genes or genes without Ensembl IDs.</span></a>
<a class="sourceLine" id="cb383-13" data-line-number="13">keep &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(gene.ids) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">duplicated</span>(gene.ids)</a>
<a class="sourceLine" id="cb383-14" data-line-number="14">sce.muraro &lt;-<span class="st"> </span>sce.muraro[keep,]</a>
<a class="sourceLine" id="cb383-15" data-line-number="15"><span class="kw">rownames</span>(sce.muraro) &lt;-<span class="st"> </span>gene.ids[keep]</a>
<a class="sourceLine" id="cb383-16" data-line-number="16"></a>
<a class="sourceLine" id="cb383-17" data-line-number="17"><span class="co">#--- quality-control ---#</span></a>
<a class="sourceLine" id="cb383-18" data-line-number="18"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb383-19" data-line-number="19">stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce.muraro)</a>
<a class="sourceLine" id="cb383-20" data-line-number="20">qc &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(stats, <span class="dt">percent_subsets=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>,</a>
<a class="sourceLine" id="cb383-21" data-line-number="21">    <span class="dt">batch=</span>sce.muraro<span class="op">$</span>donor, <span class="dt">subset=</span>sce.muraro<span class="op">$</span>donor<span class="op">!=</span><span class="st">&quot;D28&quot;</span>)</a>
<a class="sourceLine" id="cb383-22" data-line-number="22">sce.muraro &lt;-<span class="st"> </span>sce.muraro[,<span class="op">!</span>qc<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb383-23" data-line-number="23"></a>
<a class="sourceLine" id="cb383-24" data-line-number="24"><span class="co">#--- normalization ---#</span></a>
<a class="sourceLine" id="cb383-25" data-line-number="25"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb383-26" data-line-number="26"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb383-27" data-line-number="27">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.muraro)</a>
<a class="sourceLine" id="cb383-28" data-line-number="28">sce.muraro &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.muraro, <span class="dt">clusters=</span>clusters)</a>
<a class="sourceLine" id="cb383-29" data-line-number="29">sce.muraro &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce.muraro)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb384-1" data-line-number="1">sce.muraro</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 16940 2299 
## metadata(0):
## assays(2): counts logcounts
## rownames(16940): ENSG00000268895 ENSG00000121410 ... ENSG00000159840
##   ENSG00000074755
## rowData names(2): symbol chr
## colnames(2299): D28-1_1 D28-1_2 ... D30-8_93 D30-8_94
## colData names(4): label donor plate sizeFactor
## reducedDimNames(0):
## altExpNames(1): ERCC</code></pre>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb386-1" data-line-number="1"><span class="co"># Pruning out unknown or unclear labels.</span></a>
<a class="sourceLine" id="cb386-2" data-line-number="2">sce.muraro &lt;-<span class="st"> </span>sce.muraro[,<span class="op">!</span><span class="kw">is.na</span>(sce.muraro<span class="op">$</span>label) <span class="op">&amp;</span><span class="st"> </span></a>
<a class="sourceLine" id="cb386-3" data-line-number="3"><span class="st">    </span>sce.muraro<span class="op">$</span>label<span class="op">!=</span><span class="st">&quot;unclear&quot;</span>]</a>
<a class="sourceLine" id="cb386-4" data-line-number="4"><span class="kw">table</span>(sce.muraro<span class="op">$</span>label)</a></code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         217         795         442         189         239          18 
##     epsilon mesenchymal          pp 
##           3          80          96</code></pre>
<p>Our aim is to assign labels to our test dataset from <span class="citation">Segerstolpe et al. (<a href="#ref-segerstolpe2016singlecell">2016</a>)</span>.
We use the same call to <code>SingleR()</code> but with <code>de.method=&quot;wilcox&quot;</code> to identify markers via pairwise Wilcoxon ranked sum tests between labels in the reference Muraro dataset.
This re-uses the same machinery from Chapter <a href="marker-detection.html#marker-detection">11</a>; further options to fine-tune the test procedure can be passed via the <code>de.args</code> argument.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb388-1" data-line-number="1"><span class="co">#--- loading ---#</span></a>
<a class="sourceLine" id="cb388-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb388-3" data-line-number="3">sce.seger &lt;-<span class="st"> </span><span class="kw">SegerstolpePancreasData</span>()</a>
<a class="sourceLine" id="cb388-4" data-line-number="4"></a>
<a class="sourceLine" id="cb388-5" data-line-number="5"><span class="co">#--- gene-annotation ---#</span></a>
<a class="sourceLine" id="cb388-6" data-line-number="6"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb388-7" data-line-number="7">edb &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73881&quot;</span>]]</a>
<a class="sourceLine" id="cb388-8" data-line-number="8">symbols &lt;-<span class="st"> </span><span class="kw">rowData</span>(sce.seger)<span class="op">$</span>symbol</a>
<a class="sourceLine" id="cb388-9" data-line-number="9">ens.id &lt;-<span class="st"> </span><span class="kw">mapIds</span>(edb, <span class="dt">keys=</span>symbols, <span class="dt">keytype=</span><span class="st">&quot;SYMBOL&quot;</span>, <span class="dt">column=</span><span class="st">&quot;GENEID&quot;</span>)</a>
<a class="sourceLine" id="cb388-10" data-line-number="10">ens.id &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">is.na</span>(ens.id), symbols, ens.id)</a>
<a class="sourceLine" id="cb388-11" data-line-number="11"></a>
<a class="sourceLine" id="cb388-12" data-line-number="12"><span class="co"># Removing duplicated rows.</span></a>
<a class="sourceLine" id="cb388-13" data-line-number="13">keep &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">duplicated</span>(ens.id)</a>
<a class="sourceLine" id="cb388-14" data-line-number="14">sce.seger &lt;-<span class="st"> </span>sce.seger[keep,]</a>
<a class="sourceLine" id="cb388-15" data-line-number="15"><span class="kw">rownames</span>(sce.seger) &lt;-<span class="st"> </span>ens.id[keep]</a>
<a class="sourceLine" id="cb388-16" data-line-number="16"></a>
<a class="sourceLine" id="cb388-17" data-line-number="17"><span class="co">#--- sample-annotation ---#</span></a>
<a class="sourceLine" id="cb388-18" data-line-number="18">emtab.meta &lt;-<span class="st"> </span><span class="kw">colData</span>(sce.seger)[,<span class="kw">c</span>(<span class="st">&quot;cell type&quot;</span>, <span class="st">&quot;disease&quot;</span>,</a>
<a class="sourceLine" id="cb388-19" data-line-number="19">    <span class="st">&quot;individual&quot;</span>, <span class="st">&quot;single cell well quality&quot;</span>)]</a>
<a class="sourceLine" id="cb388-20" data-line-number="20"><span class="kw">colnames</span>(emtab.meta) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CellType&quot;</span>, <span class="st">&quot;Disease&quot;</span>, <span class="st">&quot;Donor&quot;</span>, <span class="st">&quot;Quality&quot;</span>)</a>
<a class="sourceLine" id="cb388-21" data-line-number="21"><span class="kw">colData</span>(sce.seger) &lt;-<span class="st"> </span>emtab.meta</a>
<a class="sourceLine" id="cb388-22" data-line-number="22"></a>
<a class="sourceLine" id="cb388-23" data-line-number="23">sce.seger<span class="op">$</span>CellType &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot; cell&quot;</span>, <span class="st">&quot;&quot;</span>, sce.seger<span class="op">$</span>CellType)</a>
<a class="sourceLine" id="cb388-24" data-line-number="24">sce.seger<span class="op">$</span>CellType &lt;-<span class="st"> </span><span class="kw">paste0</span>(</a>
<a class="sourceLine" id="cb388-25" data-line-number="25">    <span class="kw">toupper</span>(<span class="kw">substr</span>(sce.seger<span class="op">$</span>CellType, <span class="dv">1</span>, <span class="dv">1</span>)),</a>
<a class="sourceLine" id="cb388-26" data-line-number="26">    <span class="kw">substring</span>(sce.seger<span class="op">$</span>CellType, <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb388-27" data-line-number="27"></a>
<a class="sourceLine" id="cb388-28" data-line-number="28"><span class="co">#--- quality-control ---#</span></a>
<a class="sourceLine" id="cb388-29" data-line-number="29">low.qual &lt;-<span class="st"> </span>sce.seger<span class="op">$</span>Quality <span class="op">==</span><span class="st"> &quot;low quality cell&quot;</span></a>
<a class="sourceLine" id="cb388-30" data-line-number="30"></a>
<a class="sourceLine" id="cb388-31" data-line-number="31"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb388-32" data-line-number="32">stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce.seger)</a>
<a class="sourceLine" id="cb388-33" data-line-number="33">qc &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(stats, <span class="dt">percent_subsets=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>,</a>
<a class="sourceLine" id="cb388-34" data-line-number="34">    <span class="dt">batch=</span>sce.seger<span class="op">$</span>Donor,</a>
<a class="sourceLine" id="cb388-35" data-line-number="35">    <span class="dt">subset=</span><span class="op">!</span>sce.seger<span class="op">$</span>Donor <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;HP1504901&quot;</span>, <span class="st">&quot;HP1509101&quot;</span>))</a>
<a class="sourceLine" id="cb388-36" data-line-number="36"></a>
<a class="sourceLine" id="cb388-37" data-line-number="37">sce.seger &lt;-<span class="st"> </span>sce.seger[,<span class="op">!</span>(qc<span class="op">$</span>discard <span class="op">|</span><span class="st"> </span>low.qual)]</a>
<a class="sourceLine" id="cb388-38" data-line-number="38"></a>
<a class="sourceLine" id="cb388-39" data-line-number="39"><span class="co">#--- normalization ---#</span></a>
<a class="sourceLine" id="cb388-40" data-line-number="40"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb388-41" data-line-number="41">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.seger)</a>
<a class="sourceLine" id="cb388-42" data-line-number="42">sce.seger &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.seger, <span class="dt">clusters=</span>clusters)</a>
<a class="sourceLine" id="cb388-43" data-line-number="43">sce.seger &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce.seger) </a></code></pre></div>
</div>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb389-1" data-line-number="1"><span class="co"># Converting to FPKM for a more like-for-like comparison to UMI counts.</span></a>
<a class="sourceLine" id="cb389-2" data-line-number="2"><span class="co"># However, results are often still good even when this step is skipped.</span></a>
<a class="sourceLine" id="cb389-3" data-line-number="3"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb389-4" data-line-number="4">hs.db &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73881&quot;</span>]]</a>
<a class="sourceLine" id="cb389-5" data-line-number="5">hs.exons &lt;-<span class="st"> </span><span class="kw">exonsBy</span>(hs.db, <span class="dt">by=</span><span class="st">&quot;gene&quot;</span>)</a>
<a class="sourceLine" id="cb389-6" data-line-number="6">hs.exons &lt;-<span class="st"> </span><span class="kw">reduce</span>(hs.exons)</a>
<a class="sourceLine" id="cb389-7" data-line-number="7">hs.len &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">width</span>(hs.exons))</a>
<a class="sourceLine" id="cb389-8" data-line-number="8"></a>
<a class="sourceLine" id="cb389-9" data-line-number="9"><span class="kw">library</span>(scuttle)</a>
<a class="sourceLine" id="cb389-10" data-line-number="10">available &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(sce.seger), <span class="kw">names</span>(hs.len))</a>
<a class="sourceLine" id="cb389-11" data-line-number="11">fpkm.seger &lt;-<span class="st"> </span><span class="kw">calculateFPKM</span>(sce.seger[available,], hs.len[available])</a>
<a class="sourceLine" id="cb389-12" data-line-number="12"></a>
<a class="sourceLine" id="cb389-13" data-line-number="13">pred.seger &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>fpkm.seger, <span class="dt">ref=</span>sce.muraro, </a>
<a class="sourceLine" id="cb389-14" data-line-number="14">    <span class="dt">labels=</span>sce.muraro<span class="op">$</span>label, <span class="dt">de.method=</span><span class="st">&quot;wilcox&quot;</span>)</a>
<a class="sourceLine" id="cb389-15" data-line-number="15"><span class="kw">table</span>(pred.seger<span class="op">$</span>labels)</a></code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         192         892         273         106         381          18 
##     epsilon mesenchymal          pp 
##           5          52         171</code></pre>
<p>As it so happens, we are in the fortunate position where our test dataset also contains independently defined labels.
We see strong consistency between the two sets of labels (Figure <a href="cell-type-annotation.html#fig:singler-comp-pancreas">12.3</a>), indicating that our automatic annotation is comparable to that generated manually by domain experts.</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb391-1" data-line-number="1">tab &lt;-<span class="st"> </span><span class="kw">table</span>(pred.seger<span class="op">$</span>pruned.labels, sce.seger<span class="op">$</span>CellType)</a>
<a class="sourceLine" id="cb391-2" data-line-number="2"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb391-3" data-line-number="3"><span class="kw">pheatmap</span>(<span class="kw">log2</span>(tab<span class="op">+</span><span class="dv">10</span>), <span class="dt">color=</span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;blue&quot;</span>))(<span class="dv">101</span>))</a></code></pre></div>
<div class="figure"><span id="fig:singler-comp-pancreas"></span>
<img src="cell-annotation_files/figure-html/singler-comp-pancreas-1.png" alt="Heatmap of the confusion matrix between the predicted labels (rows) and the independently defined labels (columns) in the Segerstolpe dataset. The color is proportinal to the log-transformed number of cells with a given combination of labels from each set." width="672" />
<p class="caption">
Figure 12.3: Heatmap of the confusion matrix between the predicted labels (rows) and the independently defined labels (columns) in the Segerstolpe dataset. The color is proportinal to the log-transformed number of cells with a given combination of labels from each set.
</p>
</div>
<p>An interesting question is - given a single-cell reference dataset, is it better to use it directly or convert it to pseudo-bulk values?
A single-cell reference preserves the “shape” of the subpopulation in high-dimensional expression space, potentially yielding more accurate predictions when the differences between labels are subtle (or at least capturing ambiguity more accurately to avoid grossly incorrect predictions).
However, it also requires more computational work to assign each cell in the test dataset.
We refer to the <a href="https://ltla.github.io/SingleRBook/using-single-cell-references.html#pseudo-bulk-aggregation">other book</a> for more details on how to achieve a compromise between these two concerns.</p>
</div>
</div>
<div id="assigning-cell-labels-from-gene-sets" class="section level2">
<h2><span class="header-section-number">12.3</span> Assigning cell labels from gene sets</h2>
<p>A related strategy is to explicitly identify sets of marker genes that are highly expressed in each individual cell.
This does not require matching of individual cells to the expression values of the reference dataset, which is faster and more convenient when only the identities of the markers are available.
We demonstrate this approach using neuronal cell type markers derived from the <span class="citation">Zeisel et al. (<a href="#ref-zeisel2015brain">2015</a>)</span> study.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb392-1" data-line-number="1"><span class="co">#--- loading ---#</span></a>
<a class="sourceLine" id="cb392-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb392-3" data-line-number="3">sce.zeisel &lt;-<span class="st"> </span><span class="kw">ZeiselBrainData</span>()</a>
<a class="sourceLine" id="cb392-4" data-line-number="4"></a>
<a class="sourceLine" id="cb392-5" data-line-number="5"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb392-6" data-line-number="6">sce.zeisel &lt;-<span class="st"> </span><span class="kw">aggregateAcrossFeatures</span>(sce.zeisel, </a>
<a class="sourceLine" id="cb392-7" data-line-number="7">    <span class="dt">id=</span><span class="kw">sub</span>(<span class="st">&quot;_loc[0-9]+$&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">rownames</span>(sce.zeisel)))</a>
<a class="sourceLine" id="cb392-8" data-line-number="8"></a>
<a class="sourceLine" id="cb392-9" data-line-number="9"><span class="co">#--- gene-annotation ---#</span></a>
<a class="sourceLine" id="cb392-10" data-line-number="10"><span class="kw">library</span>(org.Mm.eg.db)</a>
<a class="sourceLine" id="cb392-11" data-line-number="11"><span class="kw">rowData</span>(sce.zeisel)<span class="op">$</span>Ensembl &lt;-<span class="st"> </span><span class="kw">mapIds</span>(org.Mm.eg.db, </a>
<a class="sourceLine" id="cb392-12" data-line-number="12">    <span class="dt">keys=</span><span class="kw">rownames</span>(sce.zeisel), <span class="dt">keytype=</span><span class="st">&quot;SYMBOL&quot;</span>, <span class="dt">column=</span><span class="st">&quot;ENSEMBL&quot;</span>)</a>
<a class="sourceLine" id="cb392-13" data-line-number="13"></a>
<a class="sourceLine" id="cb392-14" data-line-number="14"><span class="co">#--- quality-control ---#</span></a>
<a class="sourceLine" id="cb392-15" data-line-number="15">stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce.zeisel, <span class="dt">subsets=</span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb392-16" data-line-number="16">    <span class="dt">Mt=</span><span class="kw">rowData</span>(sce.zeisel)<span class="op">$</span>featureType<span class="op">==</span><span class="st">&quot;mito&quot;</span>))</a>
<a class="sourceLine" id="cb392-17" data-line-number="17">qc &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(stats, <span class="dt">percent_subsets=</span><span class="kw">c</span>(<span class="st">&quot;altexps_ERCC_percent&quot;</span>, </a>
<a class="sourceLine" id="cb392-18" data-line-number="18">    <span class="st">&quot;subsets_Mt_percent&quot;</span>))</a>
<a class="sourceLine" id="cb392-19" data-line-number="19">sce.zeisel &lt;-<span class="st"> </span>sce.zeisel[,<span class="op">!</span>qc<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb392-20" data-line-number="20"></a>
<a class="sourceLine" id="cb392-21" data-line-number="21"><span class="co">#--- normalization ---#</span></a>
<a class="sourceLine" id="cb392-22" data-line-number="22"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb392-23" data-line-number="23"><span class="kw">set.seed</span>(<span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb392-24" data-line-number="24">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.zeisel)</a>
<a class="sourceLine" id="cb392-25" data-line-number="25">sce.zeisel &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.zeisel, <span class="dt">cluster=</span>clusters) </a>
<a class="sourceLine" id="cb392-26" data-line-number="26">sce.zeisel &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce.zeisel)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb393-1" data-line-number="1"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb393-2" data-line-number="2">wilcox.z &lt;-<span class="st"> </span><span class="kw">pairwiseWilcox</span>(sce.zeisel, sce.zeisel<span class="op">$</span>level1class, </a>
<a class="sourceLine" id="cb393-3" data-line-number="3">    <span class="dt">lfc=</span><span class="dv">1</span>, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)</a>
<a class="sourceLine" id="cb393-4" data-line-number="4">markers.z &lt;-<span class="st"> </span><span class="kw">getTopMarkers</span>(wilcox.z<span class="op">$</span>statistics, wilcox.z<span class="op">$</span>pairs,</a>
<a class="sourceLine" id="cb393-5" data-line-number="5">    <span class="dt">pairwise=</span><span class="ot">FALSE</span>, <span class="dt">n=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb393-6" data-line-number="6"><span class="kw">lengths</span>(markers.z)</a></code></pre></div>
<pre><code>## astrocytes_ependymal    endothelial-mural         interneurons 
##                   79                   83                  118 
##            microglia     oligodendrocytes        pyramidal CA1 
##                   69                   81                  125 
##         pyramidal SS 
##                  149</code></pre>
<p>Our test dataset will be another brain scRNA-seq experiment from <span class="citation">Tasic et al. (<a href="#ref-tasic2016adult">2016</a>)</span>.</p>
<div class="sourceCode" id="cb395"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb395-1" data-line-number="1"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb395-2" data-line-number="2">sce.tasic &lt;-<span class="st"> </span><span class="kw">TasicBrainData</span>()</a>
<a class="sourceLine" id="cb395-3" data-line-number="3">sce.tasic</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 24058 1809 
## metadata(0):
## assays(1): counts
## rownames(24058): 0610005C13Rik 0610007C21Rik ... mt_X57780 tdTomato
## rowData names(0):
## colnames(1809): Calb2_tdTpositive_cell_1 Calb2_tdTpositive_cell_2 ...
##   Rbp4_CTX_250ng_2 Trib2_CTX_250ng_1
## colData names(13): sample_title mouse_line ... secondary_type
##   aibs_vignette_id
## reducedDimNames(0):
## altExpNames(1): ERCC</code></pre>
<p>We use the <em><a href="https://bioconductor.org/packages/3.12/AUCell">AUCell</a></em> package to identify marker sets that are highly expressed in each cell.
This method ranks genes by their expression values within each cell and constructs a response curve of the number of genes from each marker set that are present with increasing rank.
It then computes the area under the curve (AUC) for each marker set, quantifying the enrichment of those markers among the most highly expressed genes in that cell.
This is roughly similar to performing a Wilcoxon rank sum test between genes in and outside of the set, but involving only the top ranking genes by expression in each cell.</p>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb397-1" data-line-number="1"><span class="kw">library</span>(GSEABase)</a>
<a class="sourceLine" id="cb397-2" data-line-number="2">all.sets &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">names</span>(markers.z), <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb397-3" data-line-number="3">    <span class="kw">GeneSet</span>(markers.z[[x]], <span class="dt">setName=</span>x)        </a>
<a class="sourceLine" id="cb397-4" data-line-number="4">})</a>
<a class="sourceLine" id="cb397-5" data-line-number="5">all.sets &lt;-<span class="st"> </span><span class="kw">GeneSetCollection</span>(all.sets)</a>
<a class="sourceLine" id="cb397-6" data-line-number="6"></a>
<a class="sourceLine" id="cb397-7" data-line-number="7"><span class="kw">library</span>(AUCell)</a>
<a class="sourceLine" id="cb397-8" data-line-number="8">rankings &lt;-<span class="st"> </span><span class="kw">AUCell_buildRankings</span>(<span class="kw">counts</span>(sce.tasic),</a>
<a class="sourceLine" id="cb397-9" data-line-number="9">    <span class="dt">plotStats=</span><span class="ot">FALSE</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb397-10" data-line-number="10">cell.aucs &lt;-<span class="st"> </span><span class="kw">AUCell_calcAUC</span>(all.sets, rankings)</a>
<a class="sourceLine" id="cb397-11" data-line-number="11">results &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">assay</span>(cell.aucs))</a>
<a class="sourceLine" id="cb397-12" data-line-number="12"><span class="kw">head</span>(results)</a></code></pre></div>
<pre><code>##                           gene sets
## cells                      astrocytes_ependymal endothelial-mural interneurons
##   Calb2_tdTpositive_cell_1               0.1387           0.04264       0.5306
##   Calb2_tdTpositive_cell_2               0.1366           0.04885       0.4538
##   Calb2_tdTpositive_cell_3               0.1087           0.07270       0.3459
##   Calb2_tdTpositive_cell_4               0.1322           0.04993       0.5113
##   Calb2_tdTpositive_cell_5               0.1513           0.07161       0.4930
##   Calb2_tdTpositive_cell_6               0.1342           0.09161       0.3378
##                           gene sets
## cells                      microglia oligodendrocytes pyramidal CA1
##   Calb2_tdTpositive_cell_1   0.04845           0.1318        0.2318
##   Calb2_tdTpositive_cell_2   0.02683           0.1211        0.2063
##   Calb2_tdTpositive_cell_3   0.03583           0.1567        0.3219
##   Calb2_tdTpositive_cell_4   0.05388           0.1481        0.2547
##   Calb2_tdTpositive_cell_5   0.06656           0.1386        0.2088
##   Calb2_tdTpositive_cell_6   0.03201           0.1553        0.4011
##                           gene sets
## cells                      pyramidal SS
##   Calb2_tdTpositive_cell_1       0.3477
##   Calb2_tdTpositive_cell_2       0.2762
##   Calb2_tdTpositive_cell_3       0.5244
##   Calb2_tdTpositive_cell_4       0.3506
##   Calb2_tdTpositive_cell_5       0.3010
##   Calb2_tdTpositive_cell_6       0.5393</code></pre>
<p>We assign cell type identity to each cell in the test dataset by taking the marker set with the top AUC as the label for that cell.
Our new labels mostly agree with the original annotation from <span class="citation">Tasic et al. (<a href="#ref-tasic2016adult">2016</a>)</span>, which is encouraging.
The only exception involves misassignment of oligodendrocyte precursors to astrocytes, which may be understandable given that they are derived from a common lineage.
In the absence of prior annotation, a more general diagnostic check is to compare the assigned labels to cluster identities, under the expectation that most cells of a single cluster would have the same label (or, if multiple labels are present, they should at least represent closely related cell states).</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb399-1" data-line-number="1">new.labels &lt;-<span class="st"> </span><span class="kw">colnames</span>(results)[<span class="kw">max.col</span>(results)]</a>
<a class="sourceLine" id="cb399-2" data-line-number="2">tab &lt;-<span class="st"> </span><span class="kw">table</span>(new.labels, sce.tasic<span class="op">$</span>broad_type)</a>
<a class="sourceLine" id="cb399-3" data-line-number="3">tab</a></code></pre></div>
<pre><code>##                       
## new.labels             Astrocyte Endothelial Cell GABA-ergic Neuron
##   astrocytes_ependymal        43                2                 0
##   endothelial-mural            0               27                 0
##   interneurons                 0                0               759
##   microglia                    0                0                 0
##   oligodendrocytes             0                0                 1
##   pyramidal SS                 0                0                 1
##                       
## new.labels             Glutamatergic Neuron Microglia Oligodendrocyte
##   astrocytes_ependymal                    0         0               0
##   endothelial-mural                       0         0               0
##   interneurons                            2         0               0
##   microglia                               0        22               0
##   oligodendrocytes                        0         0              38
##   pyramidal SS                          810         0               0
##                       
## new.labels             Oligodendrocyte Precursor Cell Unclassified
##   astrocytes_ependymal                             20            4
##   endothelial-mural                                 0            2
##   interneurons                                      0           15
##   microglia                                         0            1
##   oligodendrocytes                                  2            0
##   pyramidal SS                                      0           60</code></pre>
<p>As a diagnostic measure, we examine the distribution of AUCs across cells for each label (Figure <a href="cell-type-annotation.html#fig:auc-dist">12.4</a>).
In heterogeneous populations, the distribution for each label should be bimodal with one high-scoring peak containing cells of that cell type and a low-scoring peak containing cells of other types.
The gap between these two peaks can be used to derive a threshold for whether a label is “active” for a particular cell.
(In this case, we simply take the single highest-scoring label per cell as the labels should be mutually exclusive.)
In populations where a particular cell type is expected, lack of clear bimodality for the corresponding label may indicate that its gene set is not sufficiently informative.</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb401-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">3</span>))</a>
<a class="sourceLine" id="cb401-2" data-line-number="2"><span class="kw">AUCell_exploreThresholds</span>(cell.aucs, <span class="dt">plotHist=</span><span class="ot">TRUE</span>, <span class="dt">assign=</span><span class="ot">TRUE</span>) </a></code></pre></div>
<div class="figure"><span id="fig:auc-dist"></span>
<img src="cell-annotation_files/figure-html/auc-dist-1.png" alt="Distribution of AUCs in the Tasic brain dataset for each label in the Zeisel dataset. The blue curve represents the density estimate, the red curve represents a fitted two-component mixture of normals, the pink curve represents a fitted three-component mixture, and the grey curve represents a fitted normal distribution. Vertical lines represent threshold estimates corresponding to each estimate of the distribution." width="768" />
<p class="caption">
Figure 12.4: Distribution of AUCs in the Tasic brain dataset for each label in the Zeisel dataset. The blue curve represents the density estimate, the red curve represents a fitted two-component mixture of normals, the pink curve represents a fitted three-component mixture, and the grey curve represents a fitted normal distribution. Vertical lines represent threshold estimates corresponding to each estimate of the distribution.
</p>
</div>
<p>Interpretation of the <em><a href="https://bioconductor.org/packages/3.12/AUCell">AUCell</a></em> results is most straightforward when the marker sets are mutually exclusive, as shown above for the cell type markers.
In other applications, one might consider computing AUCs for gene sets associated with signalling or metabolic pathways.
It is likely that multiple pathways will be active in any given cell, and it is tempting to use the AUCs to quantify this activity for comparison across cells.
However, such comparisons must be interpreted with much caution as the AUCs are competitive values - any increase in one pathway’s activity will naturally reduce the AUCs for all other pathways, potentially resulting in spurious differences across the population.</p>
<p>As we mentioned previously, the advantage of the <em><a href="https://bioconductor.org/packages/3.12/AUCell">AUCell</a></em> approach is that it does not require reference expression values.
This is particularly useful when dealing with gene sets derived from the literature or other qualitative forms of biological knowledge.
For example, we might instead use single-cell signatures defined from MSigDB, obtained as shown below.</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb402-1" data-line-number="1"><span class="co"># Downloading the signatures and caching them locally.</span></a>
<a class="sourceLine" id="cb402-2" data-line-number="2"><span class="kw">library</span>(BiocFileCache)</a>
<a class="sourceLine" id="cb402-3" data-line-number="3">bfc &lt;-<span class="st"> </span><span class="kw">BiocFileCache</span>(<span class="dt">ask=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb402-4" data-line-number="4">scsig.path &lt;-<span class="st"> </span><span class="kw">bfcrpath</span>(bfc, <span class="kw">file.path</span>(<span class="st">&quot;http://software.broadinstitute.org&quot;</span>,</a>
<a class="sourceLine" id="cb402-5" data-line-number="5">    <span class="st">&quot;gsea/msigdb/supplemental/scsig.all.v1.0.symbols.gmt&quot;</span>))</a>
<a class="sourceLine" id="cb402-6" data-line-number="6">scsigs &lt;-<span class="st"> </span><span class="kw">getGmt</span>(scsig.path)</a></code></pre></div>
<p>The flipside is that information on relative expression is lost when only the marker identities are used.
The net effect of ignoring expression values is difficult to predict; for example, it may reduce performance for resolving more subtle cell types, but may also improve performance if the per-cell expression was too noisy to be useful.
Performance is also highly dependent on the gene sets themselves, which may not be defined in the same context in which they are used.
For example, applying all of the MSigDB signatures on the Muraro dataset is rather disappointing (Figure <a href="cell-type-annotation.html#fig:aucell-muraro-heat">12.5</a>), while restricting to the subset of pancreas signatures is more promising.</p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb403-1" data-line-number="1">muraro.mat &lt;-<span class="st"> </span><span class="kw">counts</span>(sce.muraro)</a>
<a class="sourceLine" id="cb403-2" data-line-number="2"><span class="kw">rownames</span>(muraro.mat) &lt;-<span class="st"> </span><span class="kw">rowData</span>(sce.muraro)<span class="op">$</span>symbol</a>
<a class="sourceLine" id="cb403-3" data-line-number="3">muraro.rankings &lt;-<span class="st"> </span><span class="kw">AUCell_buildRankings</span>(muraro.mat,</a>
<a class="sourceLine" id="cb403-4" data-line-number="4">    <span class="dt">plotStats=</span><span class="ot">FALSE</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb403-5" data-line-number="5"></a>
<a class="sourceLine" id="cb403-6" data-line-number="6"><span class="co"># Applying MsigDB to the Muraro dataset, because it&#39;s human:</span></a>
<a class="sourceLine" id="cb403-7" data-line-number="7">scsig.aucs &lt;-<span class="st"> </span><span class="kw">AUCell_calcAUC</span>(scsigs, muraro.rankings)</a>
<a class="sourceLine" id="cb403-8" data-line-number="8">scsig.results &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">assay</span>(scsig.aucs))</a>
<a class="sourceLine" id="cb403-9" data-line-number="9">full.labels &lt;-<span class="st"> </span><span class="kw">colnames</span>(scsig.results)[<span class="kw">max.col</span>(scsig.results)]</a>
<a class="sourceLine" id="cb403-10" data-line-number="10">tab &lt;-<span class="st"> </span><span class="kw">table</span>(full.labels, sce.muraro<span class="op">$</span>label)</a>
<a class="sourceLine" id="cb403-11" data-line-number="11">fullheat &lt;-<span class="st"> </span><span class="kw">pheatmap</span>(<span class="kw">log10</span>(tab<span class="op">+</span><span class="dv">10</span>), <span class="dt">color=</span>viridis<span class="op">::</span><span class="kw">viridis</span>(<span class="dv">100</span>), <span class="dt">silent=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb403-12" data-line-number="12"></a>
<a class="sourceLine" id="cb403-13" data-line-number="13"><span class="co"># Restricting to the subset of Muraro-derived gene sets:</span></a>
<a class="sourceLine" id="cb403-14" data-line-number="14">scsigs.sub &lt;-<span class="st"> </span>scsigs[<span class="kw">grep</span>(<span class="st">&quot;Pancreas&quot;</span>, <span class="kw">names</span>(scsigs))]</a>
<a class="sourceLine" id="cb403-15" data-line-number="15">sub.aucs &lt;-<span class="st"> </span><span class="kw">AUCell_calcAUC</span>(scsigs.sub, muraro.rankings)</a>
<a class="sourceLine" id="cb403-16" data-line-number="16">sub.results &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">assay</span>(sub.aucs))</a>
<a class="sourceLine" id="cb403-17" data-line-number="17">sub.labels &lt;-<span class="st"> </span><span class="kw">colnames</span>(sub.results)[<span class="kw">max.col</span>(sub.results)]</a>
<a class="sourceLine" id="cb403-18" data-line-number="18">tab &lt;-<span class="st"> </span><span class="kw">table</span>(sub.labels, sce.muraro<span class="op">$</span>label)</a>
<a class="sourceLine" id="cb403-19" data-line-number="19">subheat &lt;-<span class="st"> </span><span class="kw">pheatmap</span>(<span class="kw">log10</span>(tab<span class="op">+</span><span class="dv">10</span>), <span class="dt">color=</span>viridis<span class="op">::</span><span class="kw">viridis</span>(<span class="dv">100</span>), <span class="dt">silent=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb403-20" data-line-number="20"></a>
<a class="sourceLine" id="cb403-21" data-line-number="21">gridExtra<span class="op">::</span><span class="kw">grid.arrange</span>(fullheat[[<span class="dv">4</span>]], subheat[[<span class="dv">4</span>]])</a></code></pre></div>
<div class="figure"><span id="fig:aucell-muraro-heat"></span>
<img src="cell-annotation_files/figure-html/aucell-muraro-heat-1.png" alt="Heatmaps of the log-number of cells with each combination of known labels (columns) and assigned MSigDB signatures (rows) in the Muraro data set. The signature assigned to each cell was defined as that with the highest AUC across all (top) or all pancreas-related signatures (bottom)." width="672" />
<p class="caption">
Figure 12.5: Heatmaps of the log-number of cells with each combination of known labels (columns) and assigned MSigDB signatures (rows) in the Muraro data set. The signature assigned to each cell was defined as that with the highest AUC across all (top) or all pancreas-related signatures (bottom).
</p>
</div>
</div>
<div id="assigning-cluster-labels-from-markers" class="section level2">
<h2><span class="header-section-number">12.4</span> Assigning cluster labels from markers</h2>
<p>Yet another strategy for annotation is to perform a gene set enrichment analysis on the marker genes defining each cluster.
This identifies the pathways and processes that are (relatively) active in each cluster based on upregulation of the associated genes compared to other clusters.
We demonstrate on the mouse mammary dataset from <span class="citation">Bach et al. (<a href="#ref-bach2017differentiation">2017</a>)</span>, using markers that are identified by <code>findMarkers()</code> as being upregulated at a log-fold change threshold of 1.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode" id="cb404"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb404-1" data-line-number="1"><span class="co">#--- loading ---#</span></a>
<a class="sourceLine" id="cb404-2" data-line-number="2"><span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb404-3" data-line-number="3">sce.mam &lt;-<span class="st"> </span><span class="kw">BachMammaryData</span>(<span class="dt">samples=</span><span class="st">&quot;G_1&quot;</span>)</a>
<a class="sourceLine" id="cb404-4" data-line-number="4"></a>
<a class="sourceLine" id="cb404-5" data-line-number="5"><span class="co">#--- gene-annotation ---#</span></a>
<a class="sourceLine" id="cb404-6" data-line-number="6"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb404-7" data-line-number="7"><span class="kw">rownames</span>(sce.mam) &lt;-<span class="st"> </span><span class="kw">uniquifyFeatureNames</span>(</a>
<a class="sourceLine" id="cb404-8" data-line-number="8">    <span class="kw">rowData</span>(sce.mam)<span class="op">$</span>Ensembl, <span class="kw">rowData</span>(sce.mam)<span class="op">$</span>Symbol)</a>
<a class="sourceLine" id="cb404-9" data-line-number="9"></a>
<a class="sourceLine" id="cb404-10" data-line-number="10"><span class="kw">library</span>(AnnotationHub)</a>
<a class="sourceLine" id="cb404-11" data-line-number="11">ens.mm.v97 &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73905&quot;</span>]]</a>
<a class="sourceLine" id="cb404-12" data-line-number="12"><span class="kw">rowData</span>(sce.mam)<span class="op">$</span>SEQNAME &lt;-<span class="st"> </span><span class="kw">mapIds</span>(ens.mm.v97, <span class="dt">keys=</span><span class="kw">rowData</span>(sce.mam)<span class="op">$</span>Ensembl,</a>
<a class="sourceLine" id="cb404-13" data-line-number="13">    <span class="dt">keytype=</span><span class="st">&quot;GENEID&quot;</span>, <span class="dt">column=</span><span class="st">&quot;SEQNAME&quot;</span>)</a>
<a class="sourceLine" id="cb404-14" data-line-number="14"></a>
<a class="sourceLine" id="cb404-15" data-line-number="15"><span class="co">#--- quality-control ---#</span></a>
<a class="sourceLine" id="cb404-16" data-line-number="16">is.mito &lt;-<span class="st"> </span><span class="kw">rowData</span>(sce.mam)<span class="op">$</span>SEQNAME <span class="op">==</span><span class="st"> &quot;MT&quot;</span></a>
<a class="sourceLine" id="cb404-17" data-line-number="17">stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce.mam, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span><span class="kw">which</span>(is.mito)))</a>
<a class="sourceLine" id="cb404-18" data-line-number="18">qc &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(stats, <span class="dt">percent_subsets=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>)</a>
<a class="sourceLine" id="cb404-19" data-line-number="19">sce.mam &lt;-<span class="st"> </span>sce.mam[,<span class="op">!</span>qc<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb404-20" data-line-number="20"></a>
<a class="sourceLine" id="cb404-21" data-line-number="21"><span class="co">#--- normalization ---#</span></a>
<a class="sourceLine" id="cb404-22" data-line-number="22"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb404-23" data-line-number="23"><span class="kw">set.seed</span>(<span class="dv">101000110</span>)</a>
<a class="sourceLine" id="cb404-24" data-line-number="24">clusters &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(sce.mam)</a>
<a class="sourceLine" id="cb404-25" data-line-number="25">sce.mam &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.mam, <span class="dt">clusters=</span>clusters)</a>
<a class="sourceLine" id="cb404-26" data-line-number="26">sce.mam &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce.mam)</a>
<a class="sourceLine" id="cb404-27" data-line-number="27"></a>
<a class="sourceLine" id="cb404-28" data-line-number="28"><span class="co">#--- variance-modelling ---#</span></a>
<a class="sourceLine" id="cb404-29" data-line-number="29"><span class="kw">set.seed</span>(<span class="dv">00010101</span>)</a>
<a class="sourceLine" id="cb404-30" data-line-number="30">dec.mam &lt;-<span class="st"> </span><span class="kw">modelGeneVarByPoisson</span>(sce.mam)</a>
<a class="sourceLine" id="cb404-31" data-line-number="31">top.mam &lt;-<span class="st"> </span><span class="kw">getTopHVGs</span>(dec.mam, <span class="dt">prop=</span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb404-32" data-line-number="32"></a>
<a class="sourceLine" id="cb404-33" data-line-number="33"><span class="co">#--- dimensionality-reduction ---#</span></a>
<a class="sourceLine" id="cb404-34" data-line-number="34"><span class="kw">library</span>(BiocSingular)</a>
<a class="sourceLine" id="cb404-35" data-line-number="35"><span class="kw">set.seed</span>(<span class="dv">101010011</span>)</a>
<a class="sourceLine" id="cb404-36" data-line-number="36">sce.mam &lt;-<span class="st"> </span><span class="kw">denoisePCA</span>(sce.mam, <span class="dt">technical=</span>dec.mam, <span class="dt">subset.row=</span>top.mam)</a>
<a class="sourceLine" id="cb404-37" data-line-number="37">sce.mam &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sce.mam, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</a>
<a class="sourceLine" id="cb404-38" data-line-number="38"></a>
<a class="sourceLine" id="cb404-39" data-line-number="39"><span class="co">#--- clustering ---#</span></a>
<a class="sourceLine" id="cb404-40" data-line-number="40">snn.gr &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce.mam, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>, <span class="dt">k=</span><span class="dv">25</span>)</a>
<a class="sourceLine" id="cb404-41" data-line-number="41"><span class="kw">colLabels</span>(sce.mam) &lt;-<span class="st"> </span><span class="kw">factor</span>(igraph<span class="op">::</span><span class="kw">cluster_walktrap</span>(snn.gr)<span class="op">$</span>membership)</a></code></pre></div>
</div>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb405-1" data-line-number="1">markers.mam &lt;-<span class="st"> </span><span class="kw">findMarkers</span>(sce.mam, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>, <span class="dt">lfc=</span><span class="dv">1</span>)</a></code></pre></div>
<p>As an example, we obtain annotations for the marker genes that define cluster 2.
We will use gene sets defined by the Gene Ontology (GO) project, which describe a comprehensive range of biological processes and functions.
We define our subset of relevant marker genes at a FDR of 5% and apply the <code>goana()</code> function from the <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em> package.
This performs a hypergeometric test to identify GO terms that are overrepresented in our marker subset.
(The log-fold change threshold mentioned above is useful here, as it avoids including an excessive number of genes from the overpowered nature of per-cell DE comparisons.)</p>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb406-1" data-line-number="1">chosen &lt;-<span class="st"> &quot;2&quot;</span></a>
<a class="sourceLine" id="cb406-2" data-line-number="2">cur.markers &lt;-<span class="st"> </span>markers.mam[[chosen]]</a>
<a class="sourceLine" id="cb406-3" data-line-number="3">is.de &lt;-<span class="st"> </span>cur.markers<span class="op">$</span>FDR <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span> </a>
<a class="sourceLine" id="cb406-4" data-line-number="4"><span class="kw">summary</span>(is.de)</a></code></pre></div>
<pre><code>##    Mode   FALSE    TRUE 
## logical   27819     179</code></pre>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb408-1" data-line-number="1"><span class="co"># goana() requires Entrez IDs, some of which map to multiple</span></a>
<a class="sourceLine" id="cb408-2" data-line-number="2"><span class="co"># symbols - hence the unique() in the call below.</span></a>
<a class="sourceLine" id="cb408-3" data-line-number="3"><span class="kw">library</span>(org.Mm.eg.db)</a>
<a class="sourceLine" id="cb408-4" data-line-number="4">entrez.ids &lt;-<span class="st"> </span><span class="kw">mapIds</span>(org.Mm.eg.db, <span class="dt">keys=</span><span class="kw">rownames</span>(cur.markers), </a>
<a class="sourceLine" id="cb408-5" data-line-number="5">    <span class="dt">column=</span><span class="st">&quot;ENTREZID&quot;</span>, <span class="dt">keytype=</span><span class="st">&quot;SYMBOL&quot;</span>)</a>
<a class="sourceLine" id="cb408-6" data-line-number="6"></a>
<a class="sourceLine" id="cb408-7" data-line-number="7"><span class="kw">library</span>(limma)</a>
<a class="sourceLine" id="cb408-8" data-line-number="8">go.out &lt;-<span class="st"> </span><span class="kw">goana</span>(<span class="kw">unique</span>(entrez.ids[is.de]), <span class="dt">species=</span><span class="st">&quot;Mm&quot;</span>, </a>
<a class="sourceLine" id="cb408-9" data-line-number="9">    <span class="dt">universe=</span><span class="kw">unique</span>(entrez.ids))</a>
<a class="sourceLine" id="cb408-10" data-line-number="10"></a>
<a class="sourceLine" id="cb408-11" data-line-number="11"><span class="co"># Only keeping biological process terms that are not overly general.</span></a>
<a class="sourceLine" id="cb408-12" data-line-number="12">go.out &lt;-<span class="st"> </span>go.out[<span class="kw">order</span>(go.out<span class="op">$</span>P.DE),]</a>
<a class="sourceLine" id="cb408-13" data-line-number="13">go.useful &lt;-<span class="st"> </span>go.out[go.out<span class="op">$</span>Ont<span class="op">==</span><span class="st">&quot;BP&quot;</span> <span class="op">&amp;</span><span class="st"> </span>go.out<span class="op">$</span>N <span class="op">&lt;=</span><span class="st"> </span><span class="dv">200</span>,]</a>
<a class="sourceLine" id="cb408-14" data-line-number="14"><span class="kw">head</span>(go.useful, <span class="dv">20</span>)</a></code></pre></div>
<pre><code>##                                                                      Term Ont
## GO:0006641                                 triglyceride metabolic process  BP
## GO:0006119                                      oxidative phosphorylation  BP
## GO:0006639                                 acylglycerol metabolic process  BP
## GO:0006638                                neutral lipid metabolic process  BP
## GO:0042775         mitochondrial ATP synthesis coupled electron transport  BP
## GO:0042773                       ATP synthesis coupled electron transport  BP
## GO:0022408                      negative regulation of cell-cell adhesion  BP
## GO:0046390                          ribose phosphate biosynthetic process  BP
## GO:0009152                     purine ribonucleotide biosynthetic process  BP
## GO:0035148                                                 tube formation  BP
## GO:0050729                   positive regulation of inflammatory response  BP
## GO:0009260                            ribonucleotide biosynthetic process  BP
## GO:0022904                           respiratory electron transport chain  BP
## GO:0022900                                       electron transport chain  BP
## GO:0006164                         purine nucleotide biosynthetic process  BP
## GO:0045333                                           cellular respiration  BP
## GO:0072522                purine-containing compound biosynthetic process  BP
## GO:0071404 cellular response to low-density lipoprotein particle stimulus  BP
## GO:0019432                              triglyceride biosynthetic process  BP
## GO:0046460                             neutral lipid biosynthetic process  BP
##              N DE      P.DE
## GO:0006641  96 10 1.954e-09
## GO:0006119  86  9 1.233e-08
## GO:0006639 119 10 1.598e-08
## GO:0006638 121 10 1.877e-08
## GO:0042775  51  7 8.011e-08
## GO:0042773  52  7 9.202e-08
## GO:0022408 183 11 9.806e-08
## GO:0046390 146 10 1.125e-07
## GO:0009152 130  9 4.501e-07
## GO:0035148 173 10 5.460e-07
## GO:0050729 134  9 5.819e-07
## GO:0009260 139  9 7.926e-07
## GO:0022904  71  7 8.152e-07
## GO:0022900  75  7 1.188e-06
## GO:0006164 147  9 1.267e-06
## GO:0045333 152  9 1.674e-06
## GO:0072522 152  9 1.674e-06
## GO:0071404  15  4 3.364e-06
## GO:0019432  34  5 4.344e-06
## GO:0046460  37  5 6.687e-06</code></pre>
<p>We see an enrichment for genes involved in lipid synthesis, cell adhesion and tube formation.
Given that this is a mammary gland experiment, we might guess that cluster 2 contains luminal epithelial cells responsible for milk production and secretion.
Indeed, a closer examination of the marker list indicates that this cluster upregulates milk proteins <em>Csn2</em> and <em>Csn3</em> (Figure <a href="cell-type-annotation.html#fig:violin-milk">12.6</a>).</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb410-1" data-line-number="1"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb410-2" data-line-number="2"><span class="kw">plotExpression</span>(sce.mam, <span class="dt">features=</span><span class="kw">c</span>(<span class="st">&quot;Csn2&quot;</span>, <span class="st">&quot;Csn3&quot;</span>), </a>
<a class="sourceLine" id="cb410-3" data-line-number="3">    <span class="dt">x=</span><span class="st">&quot;label&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;label&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:violin-milk"></span>
<img src="cell-annotation_files/figure-html/violin-milk-1.png" alt="Distribution of log-expression values for _Csn2_ and _Csn3_ in each cluster." width="672" />
<p class="caption">
Figure 12.6: Distribution of log-expression values for <em>Csn2</em> and <em>Csn3</em> in each cluster.
</p>
</div>
<p>Further inspection of interesting GO terms is achieved by extracting the relevant genes.
This is usually desirable to confirm that the interpretation of the annotated biological process is appropriate.
Many terms have overlapping gene sets, so a term may only be highly ranked because it shares genes with a more relevant term that represents the active pathway.</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb411-1" data-line-number="1"><span class="co"># Extract symbols for each GO term; done once.</span></a>
<a class="sourceLine" id="cb411-2" data-line-number="2">tab &lt;-<span class="st"> </span><span class="kw">select</span>(org.Mm.eg.db, <span class="dt">keytype=</span><span class="st">&quot;SYMBOL&quot;</span>, </a>
<a class="sourceLine" id="cb411-3" data-line-number="3">    <span class="dt">keys=</span><span class="kw">rownames</span>(sce.mam), <span class="dt">columns=</span><span class="st">&quot;GOALL&quot;</span>)</a>
<a class="sourceLine" id="cb411-4" data-line-number="4">by.go &lt;-<span class="st"> </span><span class="kw">split</span>(tab[,<span class="dv">1</span>], tab[,<span class="dv">2</span>])</a>
<a class="sourceLine" id="cb411-5" data-line-number="5"></a>
<a class="sourceLine" id="cb411-6" data-line-number="6"><span class="co"># Identify genes associated with an interesting term.</span></a>
<a class="sourceLine" id="cb411-7" data-line-number="7">adhesion &lt;-<span class="st"> </span><span class="kw">unique</span>(by.go[[<span class="st">&quot;GO:0022408&quot;</span>]])</a>
<a class="sourceLine" id="cb411-8" data-line-number="8"><span class="kw">head</span>(cur.markers[<span class="kw">rownames</span>(cur.markers) <span class="op">%in%</span><span class="st"> </span>adhesion,<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>], <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## DataFrame with 10 rows and 4 columns
##               Top     p.value         FDR summary.logFC
##         &lt;integer&gt;   &lt;numeric&gt;   &lt;numeric&gt;     &lt;numeric&gt;
## Spint2         11 3.28234e-34 1.37163e-31       2.39280
## Epcam          17 8.86978e-94 7.09531e-91       2.32968
## Cebpb          21 6.76957e-16 2.03800e-13       1.80192
## Cd24a          21 3.24195e-33 1.29669e-30       1.72318
## Btn1a1         24 2.16574e-13 6.12488e-11       1.26343
## Cd9            51 1.41373e-11 3.56592e-09       2.73785
## Ceacam1        52 1.66948e-38 7.79034e-36       1.56912
## Sdc4           59 9.15001e-07 1.75467e-04       1.84014
## Anxa1          68 2.58840e-06 4.76777e-04       1.29724
## Cdh1           69 1.73658e-07 3.54897e-05       1.31265</code></pre>
<p>Gene set testing of marker lists is a reliable approach for determining if pathways are up- or down-regulated between clusters.
As the top marker genes are simply DEGs, we can directly apply well-established procedures for testing gene enrichment in DEG lists (see <a href="https://bioconductor.org/packages/release/BiocViews.html#___GeneSetEnrichment">here</a> for relevant packages).
This contrasts with the <em><a href="https://bioconductor.org/packages/3.12/AUCell">AUCell</a></em> approach where scores are not easily comparable across cells.
The downside is that all conclusions are made relative to the other clusters, making it more difficult to determine cell identity if an “outgroup” is not present in the same study.</p>
</div>
<div id="computing-gene-set-activities" class="section level2">
<h2><span class="header-section-number">12.5</span> Computing gene set activities</h2>
<p>For the sake of completeness, we should mention that we can also quantify gene set activity on a per-cell level and test for differences in activity.
This inverts the standard gene set testing procedure by combining information across genes first and then testing for differences afterwards.
To avoid the pitfalls mentioned previously for the AUCs, we simply compute the average of the log-expression values across all genes in the set for each cell.
This is less sensitive to the behavior of other genes in that cell (aside from composition biases, as discussed in Chapter <a href="normalization.html#normalization">7</a>).</p>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb413-1" data-line-number="1">aggregated &lt;-<span class="st"> </span><span class="kw">sumCountsAcrossFeatures</span>(sce.mam, by.go,</a>
<a class="sourceLine" id="cb413-2" data-line-number="2">    <span class="dt">exprs_values=</span><span class="st">&quot;logcounts&quot;</span>, <span class="dt">average=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb413-3" data-line-number="3"><span class="kw">dim</span>(aggregated) <span class="co"># rows are gene sets, columns are cells</span></a></code></pre></div>
<pre><code>## [1] 22399  2772</code></pre>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb415-1" data-line-number="1">aggregated[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##               [,1]   [,2]    [,3]   [,4]   [,5]
## GO:0000002 0.34611 0.3442 0.08630 0.2811 0.2546
## GO:0000003 0.25682 0.2629 0.20721 0.1822 0.2101
## GO:0000009 0.00000 0.0000 0.00000 0.0000 0.0000
## GO:0000010 0.00000 0.0000 0.00000 0.0000 0.0000
## GO:0000012 0.36291 0.4778 0.18489 0.0000 0.3582
## GO:0000014 0.07068 0.3256 0.11094 0.3550 0.3539
## GO:0000015 0.53680 0.3006 0.33281 0.3550 0.5630
## GO:0000016 0.00000 0.0000 0.00000 0.0000 0.0000
## GO:0000018 0.26146 0.2756 0.06163 0.1353 0.1520
## GO:0000019 0.00000 0.2220 0.00000 0.1613 0.2413</code></pre>
<p>We can then identify “differential gene set activity” between clusters by looking for significant differences in the per-set averages of the relevant cells.
For example, we observe that cluster 2 has the highest average expression for the triacylglycerol biosynthesis GO term (Figure <a href="cell-type-annotation.html#fig:lipid-synth-violin">12.7</a>), consistent with the proposed identity of those cells.</p>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb417-1" data-line-number="1"><span class="kw">plotColData</span>(sce.mam, <span class="dt">y=</span><span class="kw">I</span>(aggregated[<span class="st">&quot;GO:0019432&quot;</span>,]), <span class="dt">x=</span><span class="st">&quot;label&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:lipid-synth-violin"></span>
<img src="cell-annotation_files/figure-html/lipid-synth-violin-1.png" alt="Distribution of average log-normalized expression for genes involved in triacylglycerol biosynthesis, for all cells in each cluster of the mammary gland dataset." width="672" />
<p class="caption">
Figure 12.7: Distribution of average log-normalized expression for genes involved in triacylglycerol biosynthesis, for all cells in each cluster of the mammary gland dataset.
</p>
</div>
<p>The obvious disadvantage of this approach is that not all genes in the set may exhibit the same pattern of differences.
Non-DE genes will add noise to the per-set average, “diluting” the strength of any differences compared to an analysis that focuses directly on the DE genes (Figure <a href="cell-type-annotation.html#fig:thrsp-violin">12.8</a>).
At worst, a gene set may contain subsets of DE genes that change in opposite directions, cancelling out any differences in the per-set average.
This is not uncommon for gene sets that contain both positive and negative regulators of a particular biological process or pathway.</p>
<div class="sourceCode" id="cb418"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb418-1" data-line-number="1"><span class="co"># Choose the top-ranking gene in GO:0019432.</span></a>
<a class="sourceLine" id="cb418-2" data-line-number="2"><span class="kw">plotExpression</span>(sce.mam, <span class="st">&quot;Thrsp&quot;</span>, <span class="dt">x=</span><span class="st">&quot;label&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:thrsp-violin"></span>
<img src="cell-annotation_files/figure-html/thrsp-violin-1.png" alt="Distribution of log-normalized expression values for _Thrsp_ across all cells in each cluster of the mammary gland dataset." width="672" />
<p class="caption">
Figure 12.8: Distribution of log-normalized expression values for <em>Thrsp</em> across all cells in each cluster of the mammary gland dataset.
</p>
</div>
<p>We could attempt to use the per-set averages to identify gene sets of interest via differential testing across all possible sets, e.g., with <code>findMarkers()</code>.
However, the highest ranking gene sets in this approach tend to be very small and uninteresting because - by definition - the pitfalls mentioned above are avoided when there is only one gene in the set.
This is compounded by the fact that the log-fold changes in the per-set averages are difficult to interpret.
For these reasons, we generally reserve the use of this gene set summary statistic for visualization rather than any real statistical analysis.</p>
</div>
<div id="session-info" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /home/biocbuild/bbs-3.12-bioc/R/lib/libRblas.so
LAPACK: /home/biocbuild/bbs-3.12-bioc/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=C              
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] scater_1.17.4               ggplot2_3.3.2              
 [3] limma_3.45.10               org.Mm.eg.db_3.11.4        
 [5] AUCell_1.11.0               GSEABase_1.51.1            
 [7] graph_1.67.1                annotate_1.67.1            
 [9] XML_3.99-0.5                scRNAseq_2.3.12            
[11] scran_1.17.15               bluster_0.99.1             
[13] scuttle_0.99.12             ensembldb_2.13.1           
[15] AnnotationFilter_1.13.0     GenomicFeatures_1.41.2     
[17] AnnotationDbi_1.51.3        AnnotationHub_2.21.2       
[19] BiocFileCache_1.13.1        dbplyr_1.4.4               
[21] pheatmap_1.0.12             SingleR_1.3.7              
[23] celldex_0.99.1              SingleCellExperiment_1.11.6
[25] SummarizedExperiment_1.19.6 DelayedArray_0.15.7        
[27] matrixStats_0.56.0          Matrix_1.2-18              
[29] Biobase_2.49.0              GenomicRanges_1.41.6       
[31] GenomeInfoDb_1.25.10        IRanges_2.23.10            
[33] S4Vectors_0.27.12           BiocGenerics_0.35.4        
[35] BiocStyle_2.17.0            simpleSingleCell_1.13.16   

loaded via a namespace (and not attached):
  [1] igraph_1.2.5                  lazyeval_0.2.2               
  [3] splines_4.0.2                 BiocParallel_1.23.2          
  [5] digest_0.6.25                 htmltools_0.5.0              
  [7] GO.db_3.11.4                  viridis_0.5.1                
  [9] magrittr_1.5                  memoise_1.1.0                
 [11] mixtools_1.2.0                Biostrings_2.57.2            
 [13] R.utils_2.9.2                 askpass_1.1                  
 [15] prettyunits_1.1.1             colorspace_1.4-1             
 [17] blob_1.2.1                    rappdirs_0.3.1               
 [19] xfun_0.16                     dplyr_1.0.1                  
 [21] callr_3.4.3                   crayon_1.3.4                 
 [23] RCurl_1.98-1.2                survival_3.2-3               
 [25] glue_1.4.1                    gtable_0.3.0                 
 [27] zlibbioc_1.35.0               XVector_0.29.3               
 [29] BiocSingular_1.5.0            kernlab_0.9-29               
 [31] scales_1.1.1                  DBI_1.1.0                    
 [33] edgeR_3.31.4                  Rcpp_1.0.5                   
 [35] viridisLite_0.3.0             xtable_1.8-4                 
 [37] progress_1.2.2                dqrng_0.2.1                  
 [39] bit_4.0.4                     rsvd_1.0.3                   
 [41] httr_1.4.2                    RColorBrewer_1.1-2           
 [43] ellipsis_0.3.1                pkgconfig_2.0.3              
 [45] R.methodsS3_1.8.0             farver_2.0.3                 
 [47] CodeDepends_0.6.5             locfit_1.5-9.4               
 [49] labeling_0.3                  tidyselect_1.1.0             
 [51] rlang_0.4.7                   later_1.1.0.1                
 [53] munsell_0.5.0                 BiocVersion_3.12.0           
 [55] tools_4.0.2                   generics_0.0.2               
 [57] RSQLite_2.2.0                 ExperimentHub_1.15.1         
 [59] evaluate_0.14                 stringr_1.4.0                
 [61] fastmap_1.0.1                 yaml_2.2.1                   
 [63] processx_3.4.3                knitr_1.29                   
 [65] bit64_4.0.2                   purrr_0.3.4                  
 [67] mime_0.9                      R.oo_1.23.0                  
 [69] biomaRt_2.45.2                compiler_4.0.2               
 [71] beeswarm_0.2.3                curl_4.3                     
 [73] interactiveDisplayBase_1.27.5 tibble_3.0.3                 
 [75] statmod_1.4.34                stringi_1.4.6                
 [77] highr_0.8                     ps_1.3.4                     
 [79] lattice_0.20-41               ProtGenerics_1.21.0          
 [81] vctrs_0.3.2                   pillar_1.4.6                 
 [83] lifecycle_0.2.0               BiocManager_1.30.10          
 [85] BiocNeighbors_1.7.0           cowplot_1.0.0                
 [87] data.table_1.13.0             bitops_1.0-6                 
 [89] irlba_2.3.3                   httpuv_1.5.4                 
 [91] rtracklayer_1.49.5            R6_2.4.1                     
 [93] bookdown_0.20                 promises_1.1.1               
 [95] gridExtra_2.3                 vipor_0.4.5                  
 [97] codetools_0.2-16              MASS_7.3-51.6                
 [99] assertthat_0.2.1              openssl_1.4.2                
[101] withr_2.2.0                   GenomicAlignments_1.25.3     
[103] Rsamtools_2.5.3               GenomeInfoDbData_1.2.3       
[105] hms_0.5.3                     grid_4.0.2                   
[107] rmarkdown_2.3                 DelayedMatrixStats_1.11.1    
[109] segmented_1.2-0               shiny_1.5.0                  
[111] ggbeeswarm_0.6.0             </code></pre>
</div>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-aran2019reference">
<p>Aran, D., A. P. Looney, L. Liu, E. Wu, V. Fong, A. Hsu, S. Chak, et al. 2019. “Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.” <em>Nat. Immunol.</em> 20 (2):163–72.</p>
</div>
<div id="ref-bach2017differentiation">
<p>Bach, K., S. Pensa, M. Grzelak, J. Hadfield, D. J. Adams, J. C. Marioni, and W. T. Khaled. 2017. “Differentiation dynamics of mammary epithelial cells revealed by single-cell RNA sequencing.” <em>Nat Commun</em> 8 (1):2128.</p>
</div>
<div id="ref-martens2013blueprint">
<p>Martens, J. H., and H. G. Stunnenberg. 2013. “BLUEPRINT: mapping human blood cell epigenomes.” <em>Haematologica</em> 98 (10):1487–9.</p>
</div>
<div id="ref-muraro2016singlecell">
<p>Muraro, M. J., G. Dharmadhikari, D. Grun, N. Groen, T. Dielen, E. Jansen, L. van Gurp, et al. 2016. “A Single-Cell Transcriptome Atlas of the Human Pancreas.” <em>Cell Syst</em> 3 (4):385–94.</p>
</div>
<div id="ref-segerstolpe2016singlecell">
<p>Segerstolpe, A., A. Palasantza, P. Eliasson, E. M. Andersson, A. C. Andreasson, X. Sun, S. Picelli, et al. 2016. “Single-Cell Transcriptome Profiling of Human Pancreatic Islets in Health and Type 2 Diabetes.” <em>Cell Metab.</em> 24 (4):593–607.</p>
</div>
<div id="ref-tasic2016adult">
<p>Tasic, B., V. Menon, T. N. Nguyen, T. K. Kim, T. Jarsky, Z. Yao, B. Levi, et al. 2016. “Adult mouse cortical cell taxonomy revealed by single cell transcriptomics.” <em>Nat. Neurosci.</em> 19 (2):335–46.</p>
</div>
<div id="ref-encode2012integrated">
<p>The ENCODE Project Consortium. 2012. “An integrated encyclopedia of DNA elements in the human genome.” <em>Nature</em> 489 (7414):57–74.</p>
</div>
<div id="ref-zeisel2015brain">
<p>Zeisel, A., A. B. Munoz-Manchado, S. Codeluppi, P. Lonnerberg, G. La Manno, A. Jureus, S. Marques, et al. 2015. “Brain structure. Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq.” <em>Science</em> 347 (6226):1138–42.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="marker-detection.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="integrating-datasets.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/edit/master/cell-annotation.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/commits/master/cell-annotation.Rmd",
"text": null
},
"view": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/blob/master/cell-annotation.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
